{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import shutil\n",
    "import numpy as np\n",
    "import glob\n",
    "\n",
    "from QC.utils import shell_do\n",
    "from QC.imputation import impute_data_prep\n",
    "import QC.config as config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# function definitions\n",
    "\n",
    "def make_gencall_cmds(idat_path, bpm_file, cluster_file, out_path, iaap=None):\n",
    "    \n",
    "    cmd = f'\\\n",
    "{iaap} gencall \\\n",
    "{bpm_path} \\\n",
    "{cluster_file} \\\n",
    "{out_path}/ped/ \\\n",
    "-f {idat_path}/{code} \\\n",
    "-p \\\n",
    "-t 8'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "basedir = '/data/CARD/PD/GP2/raw_genotypes'\n",
    "out_genotypes = '/data/CARD/PD/GP2/genotypes'\n",
    "cohort_path = f'{basedir}/shulman_ny'\n",
    "sample_info_path = f'{cohort_path}/sample_info'\n",
    "gtc_file_path = f'{cohort_path}/GP2_GCT_files'\n",
    "raw_idat_file_path = f'{cohort_path}/GP2_Shulman'\n",
    "ped_dir = f'{cohort_path}/ped'\n",
    "plink_dir = f'{cohort_path}/plink'\n",
    "swarm_scripts_dir = f'{cohort_path}/swarm_scripts'\n",
    "idat_dir = f'{cohort_path}/idats'\n",
    "\n",
    "# create ped and plink directories for raw geno outputs if they don't exist\n",
    "os.makedirs(ped_dir, exist_ok=True)\n",
    "os.makedirs(plink_dir, exist_ok=True)\n",
    "os.makedirs(f'{plink_dir}/indiv_samples', exist_ok=True)\n",
    "os.makedirs(swarm_scripts_dir, exist_ok=True)\n",
    "os.makedirs(idat_dir, exist_ok=True)\n",
    "\n",
    "# files\n",
    "key_file = f'{sample_info_path}/sample_key.txt'\n",
    "manifest_txt_path = f'{sample_info_path}/sample_manifest.txt'\n",
    "bpm = f'{sample_info_path}/NeuroBooster_20042459_A1.bpm'\n",
    "cluster_file = f'{sample_info_path}/NBSCluster_file_n1393_011921.egt'\n",
    "\n",
    "# phenos\n",
    "bcm_pheno_file = f'{sample_info_path}/BCM_20210201_samples.csv'\n",
    "umd_pheno_file = f'{sample_info_path}/UMD_20210201_samples.csv'\n",
    "\n",
    "# executables\n",
    "iaap = 'executables/iaap-cli-linux-x64-1.1.0-sha.80d7e5b3d9c1fdfc2e99b472a90652fd3848bbc7/iaap-cli/iaap-cli'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# some rearranging\n",
    "# !mkdir {shulman_ny_path}/sample_info\n",
    "# !cp {gtc_file_path}/Key\\ File_FINAL_Shulman_and_NY_011421.txt {sample_info_path}/sample_key.txt\n",
    "# !cp {gtc_file_path}/FINALSS_after_rerun__Shulman_and_NY_011421.csv {sample_info_path}/sample_manifest.txt\n",
    "# !cp {gtc_file_path}/NeuroBooster_20042459_A1.bpm {sample_info_path}/\n",
    "# !cp {gtc_file_path}/NBSCluster_file_n1393_011921.egt {sample_info_path}/\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create updated ID and phenotype files for later\n",
    "manifest = pd.read_csv(manifest_txt_path, header=10)\n",
    "bcm_pheno = pd.read_csv(bcm_pheno_file)\n",
    "umd_pheno = pd.read_csv(umd_pheno_file)\n",
    "pheno = bcm_pheno.append(umd_pheno)\n",
    "\n",
    "pheno['Original_clinicalID'] = pheno['Original_clinicalID'].astype(str)\n",
    "manifest['Sample_ID'] = manifest['Sample_ID'].astype(str)\n",
    "manifest['filename'] = manifest['SentrixBarcode_A'].astype(str) + '_' + manifest['SentrixPosition_A']\n",
    "pheno_out = manifest.merge(pheno, how='left', left_on='Sample_ID', right_on='Original_clinicalID')\n",
    "pheno_out['IID'] = pheno_out.SentrixBarcode_A.astype(str) + '_' + pheno_out.SentrixPosition_A.astype(str)\n",
    "pheno_out['FID'] = 0\n",
    "pheno_out['FID_new'] = 0\n",
    "pheno_out['pheno'] = 0\n",
    "pheno_out.loc[pheno_out.Phenotype == 'PD', 'pheno'] = 2\n",
    "pheno_out.loc[pheno_out.Phenotype == 'Control', 'pheno'] = 1\n",
    "pheno_out.loc[pheno_out.Phenotype == np.nan, 'pheno'] = 0\n",
    "\n",
    "pheno_out[['FID','IID', 'FID_new', 'Sample_ID']].to_csv(f'{sample_info_path}/update_ids.txt', sep='\\t', header=False, index=False)\n",
    "pheno_out[['FID_new', 'Sample_ID', 'pheno']].to_csv(f'{sample_info_path}/update_pheno.txt', sep='\\t', header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create a folder in idats for each plate in new idat_dir\n",
    "for code in manifest.SentrixBarcode_A.unique():\n",
    "    if os.path.exists(f'{idat_dir}/{code}'):\n",
    "        print(f'{idat_dir}/{code} already exists')\n",
    "    else:\n",
    "        os.mkdir(f'{idat_dir}/{code}')\n",
    "\n",
    "# copy idat intensity files to respective directories under idat_dir\n",
    "missing_idats = []\n",
    "\n",
    "for i, filename in enumerate(manifest.filename):\n",
    "    sentrix_code = manifest.SentrixBarcode_A.iloc[i]\n",
    "    grn = f'{raw_idat_file_path}/{sentrix_code}/{filename}_Grn.idat'\n",
    "    red = f'{raw_idat_file_path}/{sentrix_code}/{filename}_Red.idat'\n",
    "\n",
    "    if os.path.isfile(grn):\n",
    "        shutil.copyfile(src=grn, dst=f'{idat_dir}/{sentrix_code}/{filename}_Grn.idat')\n",
    "    else:\n",
    "        missing_idats.append(grn)\n",
    "\n",
    "    if os.path.isfile(red):\n",
    "        shutil.copyfile(src=red, dst=f'{idat_dir}/{sentrix_code}/{filename}_Red.idat')\n",
    "    else:\n",
    "        missing_idats.append(red)\n",
    "\n",
    "len(missing_idats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f'{swarm_scripts_dir}/idat_to_ped.swarm', 'w') as f:\n",
    "    \n",
    "    for code in manifest.SentrixBarcode_A.unique():\n",
    "        \n",
    "        idat_to_ped_cmd = f'\\\n",
    "{iaap} gencall \\\n",
    "{bpm} \\\n",
    "{cluster_file} \\\n",
    "{ped_dir}/ \\\n",
    "-f {idat_dir}/{code} \\\n",
    "-p \\\n",
    "-t 8'\n",
    "        \n",
    "        f.write(f'{idat_to_ped_cmd}\\n')\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12454654\n"
     ]
    }
   ],
   "source": [
    "# !swarm -f {swarm_scripts_dir}/idat_to_ped.swarm -g 32 -t 16 --time=10:00:00 --logdir swarm --gres=lscratch:20 --partition=norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "# copy map file to match name of each ped\n",
    "map_file = f'{ped_dir}/NeuroBooster_20042459_A1.map'\n",
    "for filename in manifest.filename:\n",
    "    ped = f'{ped_dir}/{filename}.ped'\n",
    "    out_map = f'{ped_dir}/{filename}.map'\n",
    "    if os.path.isfile(ped):\n",
    "        shutil.copyfile(src=map_file, dst=out_map)\n",
    "    else:\n",
    "        print(f'{ped} does not exist!')\n",
    "        print(f'{out_map} creation cancelled')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "with open(f'{swarm_scripts_dir}/make_bed.swarm', 'w') as f:\n",
    "    for filename in manifest.filename:\n",
    "        ped = f'{ped_dir}/{filename}'\n",
    "        make_bed_cmd = f'\\\n",
    "plink \\\n",
    "--file {ped} \\\n",
    "--make-bed \\\n",
    "--out {plink_dir}/indiv_samples/{filename}'\n",
    "\n",
    "        f.write(f'{make_bed_cmd}\\n')\n",
    "f.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12455471\n"
     ]
    }
   ],
   "source": [
    "# !swarm -f {swarm_scripts_dir}/make_bed.swarm -g 8 -t 8 --time=10:00:00 --logdir swarm --gres=lscratch:20 --partition=norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write plink merge command\n",
    "with open(f\"{plink_dir}/merge_bed.list\", 'w') as f:\n",
    "    for filename in manifest.filename:\n",
    "        bed = f'{plink_dir}/indiv_samples/{filename}'\n",
    "        f.write(f'{bed}\\n')\n",
    "f.close()\n",
    "\n",
    "with open(f\"{swarm_scripts_dir}/merge.swarm\", 'w') as f:\n",
    "\n",
    "    plink_merge_cmd = f'\\\n",
    "plink \\\n",
    "--merge-list {plink_dir}/merge_bed.list \\\n",
    "--update-ids {sample_info_path}/update_ids.txt \\\n",
    "--make-bed \\\n",
    "--out {plink_dir}/indiv_samples/shulman_merge'\n",
    "    f.write(f\"{plink_merge_cmd}\")\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12456657\n"
     ]
    }
   ],
   "source": [
    "# !swarm -f {swarm_scripts_dir}/merge.swarm -g 64 -t 32 --time=10:00:00 --logdir swarm --gres=lscratch:20 --partition=norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /data/CARD/PD/GP2/raw_genotypes/shulman_ny/plink/shulman.log.\n",
      "Options in effect:\n",
      "  --bfile /data/CARD/PD/GP2/raw_genotypes/shulman_ny/plink/indiv_samples/shulman_merge\n",
      "  --make-bed\n",
      "  --out /data/CARD/PD/GP2/raw_genotypes/shulman_ny/plink/shulman\n",
      "  --pheno /data/CARD/PD/GP2/raw_genotypes/shulman_ny/sample_info/update_pheno.txt\n",
      "\n",
      "1547809 MB RAM detected; reserving 773904 MB for main workspace.\n",
      "2004347 variants loaded from .bim file.\n",
      "1393 people (765 males, 601 females, 27 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "/data/CARD/PD/GP2/raw_genotypes/shulman_ny/plink/shulman.nosex .\n",
      "1285 phenotype values present after --pheno.\n",
      "Warning: Ignoring phenotypes of missing-sex samples.  If you don't want those\n",
      "phenotypes to be ignored, use the --allow-no-sex flag.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 1393 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Warning: 13952 het. haploid genotypes present (see\n",
      "/data/CARD/PD/GP2/raw_genotypes/shulman_ny/plink/shulman.hh ); many commands\n",
      "treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Total genotyping rate is 0.989111.\n",
      "2004347 variants and 1393 people pass filters and QC.\n",
      "Among remaining phenotypes, 941 are cases and 344 are controls.  (108\n",
      "phenotypes are missing.)\n",
      "--make-bed to /data/CARD/PD/GP2/raw_genotypes/shulman_ny/plink/shulman.bed +\n",
      "/data/CARD/PD/GP2/raw_genotypes/shulman_ny/plink/shulman.bim +\n",
      "/data/CARD/PD/GP2/raw_genotypes/shulman_ny/plink/shulman.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n"
     ]
    }
   ],
   "source": [
    "!plink --bfile {plink_dir}/indiv_samples/shulman_merge --pheno {sample_info_path}/update_pheno.txt --make-bed --out {plink_dir}/shulman"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "python3 ../run_qc_pipeline.py --geno /data/CARD/PD/GP2/raw_genotypes/shulman_ny/plink/shulman --ref /data/LNG/vitaled2/1kgenomes/1kg_ashkj_ref_panel_gp2_pruned --ref_labels /data/LNG/vitaled2/1kgenomes/ref_panel_ancestry.txt --out /data/CARD/PD/GP2/genotypes/shulman_ny/clean/shulman_ny\n"
     ]
    }
   ],
   "source": [
    "# run QC pipeline\n",
    "geno_path = f'{basedir}/shulman_ny/plink/shulman'\n",
    "out_dir = f'{out_genotypes}/shulman_ny/clean'\n",
    "out_path = f'{out_dir}/shulman_ny'\n",
    "os.makedirs(f'{out_dir}', exist_ok=True)\n",
    "ref_dir_path = '/data/LNG/vitaled2/1kgenomes'\n",
    "ref_panel = f'{ref_dir_path}/1kg_ashkj_ref_panel_gp2_pruned'\n",
    "ref_labels = f'{ref_dir_path}/ref_panel_ancestry.txt'\n",
    "\n",
    "\n",
    "with open(f'{swarm_scripts_dir}/run_qc_pipeline.swarm','w') as f:\n",
    "    run_pipeline = f'python3 ../run_qc_pipeline.py --geno {geno_path} --ref {ref_panel} --ref_labels {ref_labels} --out {out_path}'\n",
    "    f.write(f'{run_pipeline}\\n')\n",
    "f.close()\n",
    "!cat {swarm_scripts_dir}/run_qc_pipeline.swarm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18227274\n"
     ]
    }
   ],
   "source": [
    "!swarm -f {swarm_scripts_dir}/run_qc_pipeline.swarm -g 64 -t 32 --time=10:00:00 --logdir swarm --gres=lscratch:20 --partition=norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make new dirs for QC'd data \n",
    "cohorts = ['bcm','umd']\n",
    "\n",
    "for cohort in cohorts:\n",
    "    cohort_dir = f'{out_genotypes}/{cohort}'\n",
    "    clean_dir = f'{cohort_dir}/clean'\n",
    "    impute_dir = f'{cohort_dir}/imputed'\n",
    "    os.makedirs(cohort_dir, exist_ok=True)\n",
    "    os.makedirs(clean_dir, exist_ok=True)\n",
    "    os.makedirs(impute_dir, exist_ok=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "labels = ['EUR', 'AMR', 'SAS', 'AJ', 'AAC', 'EAS']\n",
    "\n",
    "bcm_pheno.Original_clinicalID = bcm_pheno.Original_clinicalID.astype(str) \n",
    "umd_pheno.Original_clinicalID = umd_pheno.Original_clinicalID.astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EUR\n",
      "BCM: (638, 16)\n",
      "UMD: (368, 16)\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Executing: plink --bfile /data/CARD/PD/GP2/genotypes/shulman_ny/clean/shulman_ny_EUR --keep /data/CARD/PD/GP2/genotypes/bcm/clean/bcm_EUR.samples --make-bed --out /data/CARD/PD/GP2/genotypes/bcm/clean/bcm_EUR\n",
      "Executing: plink --bfile /data/CARD/PD/GP2/genotypes/shulman_ny/clean/shulman_ny_EUR --keep /data/CARD/PD/GP2/genotypes/umd/clean/umd_EUR.samples --make-bed --out /data/CARD/PD/GP2/genotypes/umd/clean/umd_EUR\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AMR\n",
      "BCM: (77, 16)\n",
      "UMD: (7, 16)\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Executing: plink --bfile /data/CARD/PD/GP2/genotypes/shulman_ny/clean/shulman_ny_AMR --keep /data/CARD/PD/GP2/genotypes/bcm/clean/bcm_AMR.samples --make-bed --out /data/CARD/PD/GP2/genotypes/bcm/clean/bcm_AMR\n",
      "Executing: plink --bfile /data/CARD/PD/GP2/genotypes/shulman_ny/clean/shulman_ny_AMR --keep /data/CARD/PD/GP2/genotypes/umd/clean/umd_AMR.samples --make-bed --out /data/CARD/PD/GP2/genotypes/umd/clean/umd_AMR\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SAS\n",
      "BCM: (17, 16)\n",
      "UMD: (6, 16)\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Executing: plink --bfile /data/CARD/PD/GP2/genotypes/shulman_ny/clean/shulman_ny_SAS --keep /data/CARD/PD/GP2/genotypes/bcm/clean/bcm_SAS.samples --make-bed --out /data/CARD/PD/GP2/genotypes/bcm/clean/bcm_SAS\n",
      "Executing: plink --bfile /data/CARD/PD/GP2/genotypes/shulman_ny/clean/shulman_ny_SAS --keep /data/CARD/PD/GP2/genotypes/umd/clean/umd_SAS.samples --make-bed --out /data/CARD/PD/GP2/genotypes/umd/clean/umd_SAS\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AJ\n",
      "BCM: (27, 16)\n",
      "UMD: (57, 16)\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Executing: plink --bfile /data/CARD/PD/GP2/genotypes/shulman_ny/clean/shulman_ny_AJ --keep /data/CARD/PD/GP2/genotypes/bcm/clean/bcm_AJ.samples --make-bed --out /data/CARD/PD/GP2/genotypes/bcm/clean/bcm_AJ\n",
      "Executing: plink --bfile /data/CARD/PD/GP2/genotypes/shulman_ny/clean/shulman_ny_AJ --keep /data/CARD/PD/GP2/genotypes/umd/clean/umd_AJ.samples --make-bed --out /data/CARD/PD/GP2/genotypes/umd/clean/umd_AJ\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AAC\n",
      "BCM: (24, 16)\n",
      "UMD: (14, 16)\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Executing: plink --bfile /data/CARD/PD/GP2/genotypes/shulman_ny/clean/shulman_ny_AAC --keep /data/CARD/PD/GP2/genotypes/bcm/clean/bcm_AAC.samples --make-bed --out /data/CARD/PD/GP2/genotypes/bcm/clean/bcm_AAC\n",
      "Executing: plink --bfile /data/CARD/PD/GP2/genotypes/shulman_ny/clean/shulman_ny_AAC --keep /data/CARD/PD/GP2/genotypes/umd/clean/umd_AAC.samples --make-bed --out /data/CARD/PD/GP2/genotypes/umd/clean/umd_AAC\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EAS\n",
      "BCM: (7, 16)\n",
      "UMD: (4, 16)\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Executing: plink --bfile /data/CARD/PD/GP2/genotypes/shulman_ny/clean/shulman_ny_EAS --keep /data/CARD/PD/GP2/genotypes/bcm/clean/bcm_EAS.samples --make-bed --out /data/CARD/PD/GP2/genotypes/bcm/clean/bcm_EAS\n",
      "Executing: plink --bfile /data/CARD/PD/GP2/genotypes/shulman_ny/clean/shulman_ny_EAS --keep /data/CARD/PD/GP2/genotypes/umd/clean/umd_EAS.samples --make-bed --out /data/CARD/PD/GP2/genotypes/umd/clean/umd_EAS\n"
     ]
    }
   ],
   "source": [
    "# now, split up by cohort and ancestry\n",
    "\n",
    "for label in labels:\n",
    "    bcm_label_out = f'{out_genotypes}/bcm/clean/bcm_{label}'\n",
    "    umd_label_out = f'{out_genotypes}/umd/clean/umd_{label}'\n",
    "    geno_path = f'{out_genotypes}/shulman_ny/clean/shulman_ny_{label}'\n",
    "    fam_path = f'{geno_path}.fam'\n",
    "    fam = pd.read_csv(fam_path, sep='\\s+', header=None, names=['FID','IID','PAT','MAT','sex','pheno'])\n",
    "    fam.IID = fam.IID.astype(str)\n",
    "    \n",
    "    bcm_label = fam.merge(bcm_pheno, how='inner', left_on='IID', right_on='Original_clinicalID')\n",
    "    umd_label = fam.merge(umd_pheno, how='inner', left_on='IID', right_on='Original_clinicalID')\n",
    "    \n",
    "    print(label)\n",
    "    bcm_label[['FID','IID']].to_csv(f'{bcm_label_out}.samples', header=False, index=False, sep='\\t')\n",
    "    print(f'BCM: {bcm_label.shape}')\n",
    "    umd_label[['FID','IID']].to_csv(f'{umd_label_out}.samples', header=False, index=False, sep='\\t')\n",
    "    print(f'UMD: {umd_label.shape}')\n",
    "    print()\n",
    "    bcm_plink_cmd = f'plink --bfile {geno_path} --keep {bcm_label_out}.samples --make-bed --out {bcm_label_out}'\n",
    "    umd_plink_cmd = f'plink --bfile {geno_path} --keep {umd_label_out}.samples --make-bed --out {umd_label_out}'\n",
    "\n",
    "    cmds = [bcm_plink_cmd, umd_plink_cmd]\n",
    "    for cmd in cmds:\n",
    "        shell_do(cmd)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "python3 ../run_imputation_pipeline.py --geno /data/CARD/PD/GP2/raw_genotypes/shulman_ny/plink/AMR/AMR --token eyJjdHkiOiJ0ZXh0XC9wbGFpbiIsImFsZyI6IkhTMjU2In0.eyJtYWlsIjoidml0YWxlZDJAbmloLmdvdiIsImV4cGlyZSI6MTYyNjQwNjczMzc5OCwibmFtZSI6IkRhbiBWaXRhbGUiLCJhcGkiOnRydWUsInVzZXJuYW1lIjoidml0YWxlZDIifQ.hcHyBgJmcTZDEpFnb8t5gH1lfxSZQZHC4Lu9IhN0E18 --out /data/CARD/PD/GP2/genotypes/shulman_ny/imputed/AMR/\n",
      "python3 ../run_imputation_pipeline.py --geno /data/CARD/PD/GP2/raw_genotypes/shulman_ny/plink/AJ/AJ --token eyJjdHkiOiJ0ZXh0XC9wbGFpbiIsImFsZyI6IkhTMjU2In0.eyJtYWlsIjoidml0YWxlZDJAbmloLmdvdiIsImV4cGlyZSI6MTYyNjQwNjczMzc5OCwibmFtZSI6IkRhbiBWaXRhbGUiLCJhcGkiOnRydWUsInVzZXJuYW1lIjoidml0YWxlZDIifQ.hcHyBgJmcTZDEpFnb8t5gH1lfxSZQZHC4Lu9IhN0E18 --out /data/CARD/PD/GP2/genotypes/shulman_ny/imputed/AJ/\n",
      "python3 ../run_imputation_pipeline.py --geno /data/CARD/PD/GP2/raw_genotypes/shulman_ny/plink/EAS/EAS --token eyJjdHkiOiJ0ZXh0XC9wbGFpbiIsImFsZyI6IkhTMjU2In0.eyJtYWlsIjoidml0YWxlZDJAbmloLmdvdiIsImV4cGlyZSI6MTYyNjQwNjczMzc5OCwibmFtZSI6IkRhbiBWaXRhbGUiLCJhcGkiOnRydWUsInVzZXJuYW1lIjoidml0YWxlZDIifQ.hcHyBgJmcTZDEpFnb8t5gH1lfxSZQZHC4Lu9IhN0E18 --out /data/CARD/PD/GP2/genotypes/shulman_ny/imputed/EAS/\n",
      "python3 ../run_imputation_pipeline.py --geno /data/CARD/PD/GP2/raw_genotypes/shulman_ny/plink/AAC/AAC --token eyJjdHkiOiJ0ZXh0XC9wbGFpbiIsImFsZyI6IkhTMjU2In0.eyJtYWlsIjoidml0YWxlZDJAbmloLmdvdiIsImV4cGlyZSI6MTYyNjQwNjczMzc5OCwibmFtZSI6IkRhbiBWaXRhbGUiLCJhcGkiOnRydWUsInVzZXJuYW1lIjoidml0YWxlZDIifQ.hcHyBgJmcTZDEpFnb8t5gH1lfxSZQZHC4Lu9IhN0E18 --out /data/CARD/PD/GP2/genotypes/shulman_ny/imputed/AAC/\n",
      "python3 ../run_imputation_pipeline.py --geno /data/CARD/PD/GP2/raw_genotypes/shulman_ny/plink/SAS/SAS --token eyJjdHkiOiJ0ZXh0XC9wbGFpbiIsImFsZyI6IkhTMjU2In0.eyJtYWlsIjoidml0YWxlZDJAbmloLmdvdiIsImV4cGlyZSI6MTYyNjQwNjczMzc5OCwibmFtZSI6IkRhbiBWaXRhbGUiLCJhcGkiOnRydWUsInVzZXJuYW1lIjoidml0YWxlZDIifQ.hcHyBgJmcTZDEpFnb8t5gH1lfxSZQZHC4Lu9IhN0E18 --out /data/CARD/PD/GP2/genotypes/shulman_ny/imputed/SAS/\n",
      "python3 ../run_imputation_pipeline.py --geno /data/CARD/PD/GP2/raw_genotypes/shulman_ny/plink/EUR/EUR --token eyJjdHkiOiJ0ZXh0XC9wbGFpbiIsImFsZyI6IkhTMjU2In0.eyJtYWlsIjoidml0YWxlZDJAbmloLmdvdiIsImV4cGlyZSI6MTYyNjQwNjczMzc5OCwibmFtZSI6IkRhbiBWaXRhbGUiLCJhcGkiOnRydWUsInVzZXJuYW1lIjoidml0YWxlZDIifQ.hcHyBgJmcTZDEpFnb8t5gH1lfxSZQZHC4Lu9IhN0E18 --out /data/CARD/PD/GP2/genotypes/shulman_ny/imputed/EUR/\n"
     ]
    }
   ],
   "source": [
    "# run imputation\n",
    "\n",
    "# these files need to be in a different place eventually\n",
    "ref_panel='/data/vitaled2/GenoTools/ref/PASS.Variantsbravo-dbsnp-all.tab'\n",
    "check_bim_pl = '/data/vitaled2/GenoTools/ref/HRC-1000G-check-bim.pl'\n",
    "\n",
    "imputed_out_dir = f'{out_genotypes}/shulman_ny/imputed' \n",
    "os.makedirs(imputed_out_dir, exist_ok=True)\n",
    "\n",
    "# jenky method for pulling cleaned genos with ancestry labels for imputation... should figure out how to do this better later\n",
    "impute_genos_list = [x.split('.')[0] for x in glob.glob(f'{out_dir}/*.bed')]\n",
    "\n",
    "impute_labels_list = [x.split('/')[-1].replace('.bed','').split('_')[-1] for x in impute_genos_list]\n",
    "\n",
    "with open(f'{swarm_scripts_dir}/run_imputation_pipeline.swarm','w') as f:\n",
    "    for geno, label in zip(impute_genos_list, impute_labels_list):\n",
    "\n",
    "        label_temp_outdir = f'{basedir}/shulman_ny/plink/{label}'\n",
    "        label_temp = f'{label_temp_outdir}/{label}'\n",
    "        label_outdir = f'{imputed_out_dir}/{label}'\n",
    "        os.makedirs(f'{label_outdir}', exist_ok=True)\n",
    "        os.makedirs(f'{label_temp_outdir}', exist_ok=True)\n",
    "        impute_data = impute_data_prep(geno, label_temp, ref_panel, check_bim_pl)\n",
    "        run_pipeline = f'python3 ../run_imputation_pipeline.py --geno {label_temp} --token {config.api_key} --out {label_outdir}/'\n",
    "        f.write(f'{run_pipeline}\\n')\n",
    "        \n",
    "f.close()\n",
    "\n",
    "!cat {swarm_scripts_dir}/run_imputation_pipeline.swarm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!swarm -f {swarm_scripts_dir}/run_imputation_pipeline.swarm -g 20 -t 16 --time=80:00:00 --logdir swarm --gres=lscratch:20 --partition=norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now, convert imputed genos to plink2\n",
    "imputed_out_dir = f'{out_genotypes}/shulman_ny/imputed'\n",
    "\n",
    "chroms = [str(x) for x in range(1,23)] + ['X']\n",
    "\n",
    "with open(f'{swarm_scripts_dir}/unzip_imputed.swarm', 'w') as f:\n",
    "    for label in labels:\n",
    "        zipdir = f'{imputed_out_dir}/{label}'\n",
    "        for chrom in chroms:\n",
    "            zipfile = f'chr_{chrom}.zip'\n",
    "            unzip_cmd = f'cd {zipdir}; unzip -P imputer {zipfile}'\n",
    "            f.write(f'{unzip_cmd}\\n')\n",
    "f.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18612759\n"
     ]
    }
   ],
   "source": [
    "!swarm -f {swarm_scripts_dir}/unzip_imputed.swarm -g 20 -t 16 --time=80:00:00 --logdir swarm --gres=lscratch:20 --partition=norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f'{swarm_scripts_dir}/make_pgen.swarm', 'w') as f:\n",
    "    for label in labels:\n",
    "        for chrom in chroms:\n",
    "\n",
    "            vcf_in = f'{out_genotypes}/shulman_ny/imputed/{label}/chr{chrom}.dose.vcf.gz'\n",
    "            pgen_out = f'{out_genotypes}/shulman_ny/imputed/{label}/chr{chrom}'\n",
    "            \n",
    "            make_pgen_cmd = f\"plink2 --vcf {vcf_in} 'dosage=HDS' --make-pgen --out {pgen_out}\"\n",
    "            \n",
    "            f.write(f'{make_pgen_cmd}\\n')\n",
    "f.close()\n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18658132\n"
     ]
    }
   ],
   "source": [
    "!swarm -f {swarm_scripts_dir}/make_pgen.swarm -g 32 -t 32 --time=10:00:00 --logdir swarm --module plink/2.3-alpha --gres=lscratch:20 --partition=norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f'{swarm_scripts_dir}/pgen_update_info.swarm', 'w') as f:\n",
    "\n",
    "    fam_path = f'{geno_path}.fam'\n",
    "    fam = pd.read_csv(fam_path, sep='\\s+', header=None, names=['FID','IID','PAT','MAT','sex','pheno'])\n",
    "    \n",
    "    for label in labels:\n",
    "        for chrom in chroms:\n",
    "            basepath = f'{out_genotypes}/shulman_ny/imputed/{label}'\n",
    "            pgen_path = f'{basepath}/chr{chrom}'\n",
    "            pgen_out = f'{basepath}/shulman_ny_{label}_chr{chrom}'\n",
    "            \n",
    "            psam = pd.read_csv(f'{pgen_path}.psam', sep='\\s+')\n",
    "            psam.rename(columns={'#IID':'IID'}, inplace=True)\n",
    "            psam.loc[:,'IID2'] = psam.IID.str.replace('0_','')\n",
    "            psam.loc[:,['IID','IID2']].to_csv(f'{pgen_path}.update_ids', sep='\\t', header=False, index=False)\n",
    "            psam_fam = psam.merge(fam, how='left', left_on='IID2', right_on='IID')\n",
    "            psam_fam[['FID', 'IID_y', 'sex']].to_csv(f'{pgen_path}.sex', sep='\\t', header=False, index=False)\n",
    "            psam_fam[['FID', 'IID_y', 'pheno']].to_csv(f'{pgen_path}.pheno', sep='\\t', header=False, index=False)\n",
    "            \n",
    "            plink_cmd1 = f\"\\\n",
    "plink2 --pfile {pgen_path} \\\n",
    "--update-ids {pgen_path}.update_ids \\\n",
    "--make-pgen \\\n",
    "--out {pgen_path}_tmp\"\n",
    "            \n",
    "            plink_cmd2 = f\"\\\n",
    "plink2 --pfile {pgen_path}_tmp \\\n",
    "--update-sex {pgen_path}.sex \\\n",
    "--pheno {pgen_path}.pheno \\\n",
    "--make-pgen \\\n",
    "--out {pgen_out}\"\n",
    "\n",
    "            \n",
    "\n",
    "            f.write(f'{plink_cmd1} && {plink_cmd2}\\n')\n",
    "f.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18706935\n"
     ]
    }
   ],
   "source": [
    "!swarm -f {swarm_scripts_dir}/pgen_update_info.swarm -g 32 -t 32 --time=10:00:00 --logdir swarm --module plink/2.3-alpha --gres=lscratch:20 --partition=norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "with open(f'{swarm_scripts_dir}/split_cohort_pgen.swarm', 'w') as f: \n",
    "    for label in labels:\n",
    "        basepath = f'{out_genotypes}/shulman_ny/imputed/{label}'\n",
    "        for chrom in chroms:\n",
    "\n",
    "            bcm_label_path = f'{out_genotypes}/bcm/imputed/{label}'\n",
    "            umd_label_path = f'{out_genotypes}/umd/imputed/{label}'\n",
    "            bcm_label_out = f'{bcm_label_path}/bcm_{label}_chr{chrom}'\n",
    "            umd_label_out = f'{umd_label_path}/umd_{label}_chr{chrom}'\n",
    "\n",
    "            pgen_path = f'{basepath}/shulman_ny_{label}_chr{chrom}'\n",
    "            psam = pd.read_csv(f'{pgen_path}.psam', sep='\\s+')\n",
    "            psam.rename(columns={'#IID':'IID'}, inplace=True)\n",
    "            psam.IID = psam.IID.astype(str)\n",
    "\n",
    "            os.makedirs(bcm_label_path, exist_ok=True)\n",
    "            os.makedirs(umd_label_path, exist_ok=True)\n",
    "\n",
    "            bcm_label = psam.merge(bcm_pheno, how='inner', left_on='IID', right_on='Original_clinicalID')\n",
    "            umd_label = psam.merge(umd_pheno, how='inner', left_on='IID', right_on='Original_clinicalID')\n",
    "\n",
    "\n",
    "            bcm_label[['IID']].to_csv(f'{bcm_label_out}.samples', header=False, index=False, sep='\\t')\n",
    "            umd_label[['IID']].to_csv(f'{umd_label_out}.samples', header=False, index=False, sep='\\t')\n",
    "            bcm_plink_cmd = f'plink2 --pfile {pgen_path} --keep {bcm_label_out}.samples --make-pgen --out {bcm_label_out}'\n",
    "            umd_plink_cmd = f'plink2 --pfile {pgen_path} --keep {umd_label_out}.samples --make-pgen --out {umd_label_out}'\n",
    "            \n",
    "            cmds = [bcm_plink_cmd, umd_plink_cmd]\n",
    "            \n",
    "            for cmd in cmds:\n",
    "                f.write(f'{cmd}\\n')\n",
    "f.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18769913\n"
     ]
    }
   ],
   "source": [
    "!swarm -f {swarm_scripts_dir}/split_cohort_pgen.swarm -g 8 -t 8 --time=10:00:00 --logdir swarm --module plink/2.3-alpha --gres=lscratch:20 --partition=norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EUR\n",
      "BCM:\n",
      "13863122\n",
      "UMD:\n",
      "13863122\n",
      "\n",
      "AMR\n",
      "BCM:\n",
      "15331055\n",
      "UMD:\n",
      "15331055\n",
      "\n",
      "SAS\n",
      "BCM:\n",
      "9794453\n",
      "UMD:\n",
      "9794453\n",
      "\n",
      "AJ\n",
      "BCM:\n",
      "13762228\n",
      "UMD:\n",
      "13762228\n",
      "\n",
      "AAC\n",
      "BCM:\n",
      "15331055\n",
      "UMD:\n",
      "15331055\n",
      "\n",
      "EAS\n",
      "BCM:\n",
      "4993566\n",
      "UMD:\n",
      "4993566\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for label in labels:\n",
    "    print(label)\n",
    "    print('BCM:')\n",
    "    !cat {out_genotypes}/bcm/imputed/{label}/bcm_{label}_chrX.pvar | wc -l\n",
    "    print('UMD:')\n",
    "    !cat {out_genotypes}/umd/imputed/{label}/umd_{label}_chrX.pvar | wc -l\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data/CARD/PD/GP2/genotypes/bcm/imputed/EUR/\n",
      "/data/CARD/PD/GP2/genotypes/bcm/imputed/AMR/\n",
      "/data/CARD/PD/GP2/genotypes/bcm/imputed/SAS/\n",
      "/data/CARD/PD/GP2/genotypes/bcm/imputed/AJ/\n",
      "/data/CARD/PD/GP2/genotypes/bcm/imputed/AAC/\n",
      "/data/CARD/PD/GP2/genotypes/bcm/imputed/EAS/\n"
     ]
    }
   ],
   "source": [
    "for label in labels:\n",
    "    !echo {out_genotypes}/bcm/imputed/{label}/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python/3.7",
   "language": "python",
   "name": "py3.7"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
