{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7caa16e9-69fd-45d3-ba36-ce3ee96f726a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import shutil\n",
    "import numpy as np\n",
    "import glob\n",
    "import subprocess\n",
    "\n",
    "from QC.utils import shell_do\n",
    "import QC.config as config\n",
    "from QC.imputation import impute_data_prep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7be1a4a1-13e6-4a64-afba-01b09b8a5d19",
   "metadata": {},
   "outputs": [],
   "source": [
    "basedir = '/data/CARD/PD/GP2/raw_genotypes'\n",
    "out_genotypes = '/data/CARD/PD/GP2/genotypes'\n",
    "cohort_path = f'{basedir}/coriell'\n",
    "sample_info_path = f'{cohort_path}/sample_info'\n",
    "ilmn_files_path = '/data/CARD/PD/GP2/ilmn_files'\n",
    "bpm = f'{ilmn_files_path}/NeuroBooster_20042459_A1.bpm'\n",
    "egt = f'{ilmn_files_path}/NBSCluster_file_n1393_011921.egt'\n",
    "# key_file = f'{sample_info_path}/KEY_FILE_GP2_Coriell_P1_25_042621_B1.txt'\n",
    "key_file = f'{sample_info_path}/coriell_total_key.txt'\n",
    "pheno_file = f'{sample_info_path}/CORIELL_20210201_clinical.csv'\n",
    "\n",
    "ped_dir = f'{cohort_path}/ped'\n",
    "bed_dir = f'{cohort_path}/bed'\n",
    "swarm_scripts_dir = f'{cohort_path}/swarm_scripts'\n",
    "idat_dir = f'{cohort_path}/idats'\n",
    "raw_idat_file_path = f'{cohort_path}/idats/GP2_coriell_idats'\n",
    "\n",
    "# create ped and plink directories for raw geno outputs if they don't exist\n",
    "os.makedirs(ped_dir, exist_ok=True)\n",
    "os.makedirs(bed_dir, exist_ok=True)\n",
    "os.makedirs(f'{bed_dir}/indiv_samples', exist_ok=True)\n",
    "os.makedirs(swarm_scripts_dir, exist_ok=True)\n",
    "os.makedirs(idat_dir, exist_ok=True)\n",
    "\n",
    "iaap = f'{ilmn_files_path}/iaap-cli/iaap-cli'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "ae925bbc-47d9-4619-8696-bf6e70e3f639",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>SentrixBarcode_A</th>\n",
       "      <th>SentrixPosition_A</th>\n",
       "      <th>filename</th>\n",
       "      <th>GP2SampleID</th>\n",
       "      <th>Original_clinicalID</th>\n",
       "      <th>Original_genotypeID</th>\n",
       "      <th>Sample_submitter</th>\n",
       "      <th>Cohort_name</th>\n",
       "      <th>Country_origin</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>Race_reported</th>\n",
       "      <th>IID</th>\n",
       "      <th>FID</th>\n",
       "      <th>FID_new</th>\n",
       "      <th>pheno</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ND06487</td>\n",
       "      <td>204701860033</td>\n",
       "      <td>R01C01</td>\n",
       "      <td>204701860033_R01C01</td>\n",
       "      <td>CORIELL_USA_001831</td>\n",
       "      <td>ND06487</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Blauwendraat</td>\n",
       "      <td>CORIELL</td>\n",
       "      <td>USA</td>\n",
       "      <td>Female</td>\n",
       "      <td>PD</td>\n",
       "      <td>Black or African American</td>\n",
       "      <td>204701860033_R01C01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ND06726</td>\n",
       "      <td>204701860129</td>\n",
       "      <td>R03C01</td>\n",
       "      <td>204701860129_R03C01</td>\n",
       "      <td>CORIELL_USA_001882</td>\n",
       "      <td>ND06726</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Blauwendraat</td>\n",
       "      <td>CORIELL</td>\n",
       "      <td>USA</td>\n",
       "      <td>Female</td>\n",
       "      <td>PD</td>\n",
       "      <td>Black or African American</td>\n",
       "      <td>204701860129_R03C01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ND06940</td>\n",
       "      <td>204701860129</td>\n",
       "      <td>R04C01</td>\n",
       "      <td>204701860129_R04C01</td>\n",
       "      <td>CORIELL_USA_001943</td>\n",
       "      <td>ND06940</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Blauwendraat</td>\n",
       "      <td>CORIELL</td>\n",
       "      <td>USA</td>\n",
       "      <td>Male</td>\n",
       "      <td>PD</td>\n",
       "      <td>Black or African American</td>\n",
       "      <td>204701860129_R04C01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ND06944</td>\n",
       "      <td>204701860129</td>\n",
       "      <td>R05C01</td>\n",
       "      <td>204701860129_R05C01</td>\n",
       "      <td>CORIELL_USA_001947</td>\n",
       "      <td>ND06944</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Blauwendraat</td>\n",
       "      <td>CORIELL</td>\n",
       "      <td>USA</td>\n",
       "      <td>Female</td>\n",
       "      <td>PD</td>\n",
       "      <td>Black or African American</td>\n",
       "      <td>204701860129_R05C01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ND04086</td>\n",
       "      <td>204701860129</td>\n",
       "      <td>R08C01</td>\n",
       "      <td>204701860129_R08C01</td>\n",
       "      <td>CORIELL_USA_001082</td>\n",
       "      <td>ND04086</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Blauwendraat</td>\n",
       "      <td>CORIELL</td>\n",
       "      <td>USA</td>\n",
       "      <td>Male</td>\n",
       "      <td>Control</td>\n",
       "      <td>Black or African American</td>\n",
       "      <td>204701860129_R08C01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3930</th>\n",
       "      <td>ND21902</td>\n",
       "      <td>205275450048</td>\n",
       "      <td>R07C01</td>\n",
       "      <td>205275450048_R07C01</td>\n",
       "      <td>CORIELL_USA_004993</td>\n",
       "      <td>ND21902</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Blauwendraat</td>\n",
       "      <td>CORIELL</td>\n",
       "      <td>USA</td>\n",
       "      <td>Female</td>\n",
       "      <td>Control</td>\n",
       "      <td>Black or African American</td>\n",
       "      <td>205275450048_R07C01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3973</th>\n",
       "      <td>ND22772</td>\n",
       "      <td>204842400153</td>\n",
       "      <td>R02C01</td>\n",
       "      <td>204842400153_R02C01</td>\n",
       "      <td>CORIELL_USA_005243</td>\n",
       "      <td>ND22772</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Blauwendraat</td>\n",
       "      <td>CORIELL</td>\n",
       "      <td>USA</td>\n",
       "      <td>Male</td>\n",
       "      <td>PD</td>\n",
       "      <td>Black or African American</td>\n",
       "      <td>204842400153_R02C01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3976</th>\n",
       "      <td>ND23069</td>\n",
       "      <td>204842400153</td>\n",
       "      <td>R05C01</td>\n",
       "      <td>204842400153_R05C01</td>\n",
       "      <td>CORIELL_USA_005320</td>\n",
       "      <td>ND23069</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Blauwendraat</td>\n",
       "      <td>CORIELL</td>\n",
       "      <td>USA</td>\n",
       "      <td>Male</td>\n",
       "      <td>Control</td>\n",
       "      <td>Black or African American</td>\n",
       "      <td>204842400153_R05C01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3991</th>\n",
       "      <td>ND22181</td>\n",
       "      <td>204842400171</td>\n",
       "      <td>R04C01</td>\n",
       "      <td>204842400171_R04C01</td>\n",
       "      <td>CORIELL_USA_005103</td>\n",
       "      <td>ND22181</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Blauwendraat</td>\n",
       "      <td>CORIELL</td>\n",
       "      <td>USA</td>\n",
       "      <td>Female</td>\n",
       "      <td>PD</td>\n",
       "      <td>Black or African American</td>\n",
       "      <td>204842400171_R04C01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4007</th>\n",
       "      <td>ND23373</td>\n",
       "      <td>204825020028</td>\n",
       "      <td>R04C01</td>\n",
       "      <td>204825020028_R04C01</td>\n",
       "      <td>CORIELL_USA_005407</td>\n",
       "      <td>ND23373</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Blauwendraat</td>\n",
       "      <td>CORIELL</td>\n",
       "      <td>USA</td>\n",
       "      <td>Male</td>\n",
       "      <td>PD</td>\n",
       "      <td>Black or African American</td>\n",
       "      <td>204825020028_R04C01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>127 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Sample_ID  SentrixBarcode_A SentrixPosition_A             filename  \\\n",
       "8      ND06487      204701860033            R01C01  204701860033_R01C01   \n",
       "18     ND06726      204701860129            R03C01  204701860129_R03C01   \n",
       "19     ND06940      204701860129            R04C01  204701860129_R04C01   \n",
       "20     ND06944      204701860129            R05C01  204701860129_R05C01   \n",
       "23     ND04086      204701860129            R08C01  204701860129_R08C01   \n",
       "...        ...               ...               ...                  ...   \n",
       "3930   ND21902      205275450048            R07C01  205275450048_R07C01   \n",
       "3973   ND22772      204842400153            R02C01  204842400153_R02C01   \n",
       "3976   ND23069      204842400153            R05C01  204842400153_R05C01   \n",
       "3991   ND22181      204842400171            R04C01  204842400171_R04C01   \n",
       "4007   ND23373      204825020028            R04C01  204825020028_R04C01   \n",
       "\n",
       "             GP2SampleID Original_clinicalID Original_genotypeID  \\\n",
       "8     CORIELL_USA_001831             ND06487                 NaN   \n",
       "18    CORIELL_USA_001882             ND06726                 NaN   \n",
       "19    CORIELL_USA_001943             ND06940                 NaN   \n",
       "20    CORIELL_USA_001947             ND06944                 NaN   \n",
       "23    CORIELL_USA_001082             ND04086                 NaN   \n",
       "...                  ...                 ...                 ...   \n",
       "3930  CORIELL_USA_004993             ND21902                 NaN   \n",
       "3973  CORIELL_USA_005243             ND22772                 NaN   \n",
       "3976  CORIELL_USA_005320             ND23069                 NaN   \n",
       "3991  CORIELL_USA_005103             ND22181                 NaN   \n",
       "4007  CORIELL_USA_005407             ND23373                 NaN   \n",
       "\n",
       "     Sample_submitter Cohort_name Country_origin     Sex Phenotype  \\\n",
       "8        Blauwendraat     CORIELL            USA  Female        PD   \n",
       "18       Blauwendraat     CORIELL            USA  Female        PD   \n",
       "19       Blauwendraat     CORIELL            USA    Male        PD   \n",
       "20       Blauwendraat     CORIELL            USA  Female        PD   \n",
       "23       Blauwendraat     CORIELL            USA    Male   Control   \n",
       "...               ...         ...            ...     ...       ...   \n",
       "3930     Blauwendraat     CORIELL            USA  Female   Control   \n",
       "3973     Blauwendraat     CORIELL            USA    Male        PD   \n",
       "3976     Blauwendraat     CORIELL            USA    Male   Control   \n",
       "3991     Blauwendraat     CORIELL            USA  Female        PD   \n",
       "4007     Blauwendraat     CORIELL            USA    Male        PD   \n",
       "\n",
       "                  Race_reported                  IID  FID  FID_new  pheno  \n",
       "8     Black or African American  204701860033_R01C01    0        0      2  \n",
       "18    Black or African American  204701860129_R03C01    0        0      2  \n",
       "19    Black or African American  204701860129_R04C01    0        0      2  \n",
       "20    Black or African American  204701860129_R05C01    0        0      2  \n",
       "23    Black or African American  204701860129_R08C01    0        0      1  \n",
       "...                         ...                  ...  ...      ...    ...  \n",
       "3930  Black or African American  205275450048_R07C01    0        0      1  \n",
       "3973  Black or African American  204842400153_R02C01    0        0      2  \n",
       "3976  Black or African American  204842400153_R05C01    0        0      1  \n",
       "3991  Black or African American  204842400171_R04C01    0        0      2  \n",
       "4007  Black or African American  204825020028_R04C01    0        0      2  \n",
       "\n",
       "[127 rows x 17 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# fix sidransky ids\n",
    "# om = pd.read_csv(f'{sample_info_path}/GP2_Sidransky_GD_cohort_original_manifest.csv')\n",
    "# om\n",
    "# ids = pd.read_csv(f'{sample_info_path}/GP2_BATCH3_Coriell_P26_43_Sidransky_772021.csv')\n",
    "# for x in ids.Sample_ID.str.replace('\\W', '_',regex=True):\n",
    "#     print(x)\n",
    "# for x in pheno.Original_clinicalID:\n",
    "#     print(x)\n",
    "# pheno[pheno.Race_reported=='Black or African American']\n",
    "# pheno\n",
    "pheno_out[pheno_out.Race_reported=='Black or African American']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "5b99d7ac-100c-48f1-b014-258c4c1de653",
   "metadata": {},
   "outputs": [],
   "source": [
    "pheno = pd.read_csv(pheno_file)\n",
    "key = pd.read_csv(key_file, sep='\\t')\n",
    "\n",
    "\n",
    "pheno['Original_clinicalID'] = pheno['Original_clinicalID'].astype(str)\n",
    "key['Sample_ID'] = key['Sample_ID'].astype(str)\n",
    "key['filename'] = key['SentrixBarcode_A'].astype(str) + '_' + key['SentrixPosition_A']\n",
    "pheno_out = key.merge(pheno, how='inner', left_on='Sample_ID', right_on='Original_clinicalID')\n",
    "pheno_out['IID'] = pheno_out.SentrixBarcode_A.astype(str) + '_' + pheno_out.SentrixPosition_A.astype(str)\n",
    "pheno_out['FID'] = 0\n",
    "pheno_out['FID_new'] = 0\n",
    "pheno_out['pheno'] = 0\n",
    "pheno_out.loc[pheno_out.Phenotype == 'PD', 'pheno'] = 2\n",
    "pheno_out.loc[pheno_out.Phenotype == 'Control', 'pheno'] = 1\n",
    "pheno_out.loc[pheno_out.Phenotype == np.nan, 'pheno'] = 0\n",
    "\n",
    "pheno_out[['FID','IID', 'FID_new', 'Sample_ID']].to_csv(f'{sample_info_path}/update_ids.txt', sep='\\t', header=False, index=False)\n",
    "pheno_out[['FID_new', 'Sample_ID', 'pheno']].to_csv(f'{sample_info_path}/update_pheno.txt', sep='\\t', header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "4a32bd56-d433-4dd0-af29-122891b6fe4b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>SentrixBarcode_A</th>\n",
       "      <th>SentrixPosition_A</th>\n",
       "      <th>filename</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ND06295</td>\n",
       "      <td>204701860088</td>\n",
       "      <td>R01C01</td>\n",
       "      <td>204701860088_R01C01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ND06869</td>\n",
       "      <td>204701860088</td>\n",
       "      <td>R02C01</td>\n",
       "      <td>204701860088_R02C01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ND06634</td>\n",
       "      <td>204701860088</td>\n",
       "      <td>R03C01</td>\n",
       "      <td>204701860088_R03C01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ND06354</td>\n",
       "      <td>204701860088</td>\n",
       "      <td>R04C01</td>\n",
       "      <td>204701860088_R04C01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ND03497</td>\n",
       "      <td>204701860088</td>\n",
       "      <td>R05C01</td>\n",
       "      <td>204701860088_R05C01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4311</th>\n",
       "      <td>SidranskyGD_103</td>\n",
       "      <td>205275450145</td>\n",
       "      <td>R04C01</td>\n",
       "      <td>205275450145_R04C01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4312</th>\n",
       "      <td>SidranskyGD_105</td>\n",
       "      <td>205275450145</td>\n",
       "      <td>R05C01</td>\n",
       "      <td>205275450145_R05C01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4313</th>\n",
       "      <td>SidranskyGD_1002</td>\n",
       "      <td>205275450145</td>\n",
       "      <td>R06C01</td>\n",
       "      <td>205275450145_R06C01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4314</th>\n",
       "      <td>SidranskyGD_1005</td>\n",
       "      <td>205275450145</td>\n",
       "      <td>R07C01</td>\n",
       "      <td>205275450145_R07C01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4315</th>\n",
       "      <td>SidranskyGD_1008A</td>\n",
       "      <td>205275450145</td>\n",
       "      <td>R08C01</td>\n",
       "      <td>205275450145_R08C01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4316 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Sample_ID  SentrixBarcode_A SentrixPosition_A  \\\n",
       "0               ND06295      204701860088            R01C01   \n",
       "1               ND06869      204701860088            R02C01   \n",
       "2               ND06634      204701860088            R03C01   \n",
       "3               ND06354      204701860088            R04C01   \n",
       "4               ND03497      204701860088            R05C01   \n",
       "...                 ...               ...               ...   \n",
       "4311    SidranskyGD_103      205275450145            R04C01   \n",
       "4312    SidranskyGD_105      205275450145            R05C01   \n",
       "4313   SidranskyGD_1002      205275450145            R06C01   \n",
       "4314   SidranskyGD_1005      205275450145            R07C01   \n",
       "4315  SidranskyGD_1008A      205275450145            R08C01   \n",
       "\n",
       "                 filename  \n",
       "0     204701860088_R01C01  \n",
       "1     204701860088_R02C01  \n",
       "2     204701860088_R03C01  \n",
       "3     204701860088_R04C01  \n",
       "4     204701860088_R05C01  \n",
       "...                   ...  \n",
       "4311  205275450145_R04C01  \n",
       "4312  205275450145_R05C01  \n",
       "4313  205275450145_R06C01  \n",
       "4314  205275450145_R07C01  \n",
       "4315  205275450145_R08C01  \n",
       "\n",
       "[4316 rows x 4 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "key"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "12fe3419-780d-4e42-bdb7-bc96274daca1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create a folder in idats for each plate in new idat_dir\n",
    "for code in key.SentrixBarcode_A.unique():\n",
    "    os.makedirs(f'{idat_dir}/{code}', exist_ok=True)\n",
    "\n",
    "# copy idat intensity files to respective directories under idat_dir\n",
    "missing_idats = []\n",
    "\n",
    "for i, filename in enumerate(key.filename):\n",
    "    sentrix_code = key.SentrixBarcode_A.iloc[i]\n",
    "    grn = f'{raw_idat_file_path}/{filename}_Grn.idat'\n",
    "    red = f'{raw_idat_file_path}/{filename}_Red.idat'\n",
    "\n",
    "    if os.path.isfile(grn):\n",
    "        shutil.copyfile(src=grn, dst=f'{idat_dir}/{sentrix_code}/{filename}_Grn.idat')\n",
    "    else:\n",
    "        missing_idats.append(grn)\n",
    "\n",
    "    if os.path.isfile(red):\n",
    "        shutil.copyfile(src=red, dst=f'{idat_dir}/{sentrix_code}/{filename}_Red.idat')\n",
    "    else:\n",
    "        missing_idats.append(red)\n",
    "\n",
    "len(missing_idats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2892b657-7635-450a-8c05-9f89b931a503",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "6e0a464c-2501-42d7-8a0d-61b6f2171a7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f'{swarm_scripts_dir}/idat_to_ped.swarm', 'w') as f:\n",
    "    \n",
    "    for code in key.SentrixBarcode_A.unique():\n",
    "        \n",
    "        idat_to_ped_cmd = f'\\\n",
    "{iaap} gencall \\\n",
    "{bpm} \\\n",
    "{egt} \\\n",
    "{ped_dir}/ \\\n",
    "-f {idat_dir}/{code} \\\n",
    "-p \\\n",
    "-t 8'\n",
    "        \n",
    "        f.write(f'{idat_to_ped_cmd}\\n')\n",
    "f.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "8fac1a59-aca2-4933-879f-38b4ccb12b34",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19399969\n"
     ]
    }
   ],
   "source": [
    "!swarm -f {swarm_scripts_dir}/idat_to_ped.swarm -g 32 -t 16 --time=10:00:00 --logdir {swarm_scripts_dir}/logs --gres=lscratch:20 --partition=norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "3a6fa706-dba5-4a34-aa64-7761ebeadf1d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data/CARD/PD/GP2/raw_genotypes/coriell/ped/205275450156_R02C01.ped does not exist!\n",
      "/data/CARD/PD/GP2/raw_genotypes/coriell/ped/205275450156_R02C01.map creation cancelled\n"
     ]
    }
   ],
   "source": [
    "# copy map file to match name of each ped\n",
    "map_file = f'{ped_dir}/NeuroBooster_20042459_A1.map'\n",
    "for filename in key.filename:\n",
    "    ped = f'{ped_dir}/{filename}.ped'\n",
    "    out_map = f'{ped_dir}/{filename}.map'\n",
    "    if os.path.isfile(ped):\n",
    "        shutil.copyfile(src=map_file, dst=out_map)\n",
    "    else:\n",
    "        print(f'{ped} does not exist!')\n",
    "        print(f'{out_map} creation cancelled')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "bbf10e55-44e2-43c9-82b9-10ec7c0de9b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "with open(f'{swarm_scripts_dir}/make_bed.swarm', 'w') as f:\n",
    "    for filename in key.filename:\n",
    "        ped = f'{ped_dir}/{filename}'\n",
    "        make_bed_cmd = f'\\\n",
    "plink \\\n",
    "--file {ped} \\\n",
    "--make-bed \\\n",
    "--out {bed_dir}/indiv_samples/{filename}'\n",
    "\n",
    "        f.write(f'{make_bed_cmd}\\n')\n",
    "f.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "fa5be5ce-2359-4af9-b24b-22353829b319",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19401925\n"
     ]
    }
   ],
   "source": [
    "!swarm -f {swarm_scripts_dir}/make_bed.swarm -g 64 -t 32 --time=10:00:00 --logdir swarm --gres=lscratch:20 --partition=norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "f2bb6d75-b3c1-4e52-a110-1674825b6337",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data/CARD/PD/GP2/raw_genotypes/coriell/bed/indiv_samples/205275450156_R02C01 does not exist!!!\n"
     ]
    }
   ],
   "source": [
    "# write plink merge command\n",
    "with open(f\"{bed_dir}/merge_bed.list\", 'w') as f:\n",
    "    for filename in key.filename:\n",
    "        bed = f'{bed_dir}/indiv_samples/{filename}'\n",
    "        if os.path.isfile(f'{bed}.bed'):\n",
    "            f.write(f'{bed}\\n')\n",
    "        else:\n",
    "            print(f'{bed} does not exist!!!')\n",
    "f.close()\n",
    "\n",
    "with open(f\"{swarm_scripts_dir}/merge.swarm\", 'w') as f:\n",
    "\n",
    "    plink_merge_cmd = f'\\\n",
    "plink \\\n",
    "--merge-list {bed_dir}/merge_bed.list \\\n",
    "--update-ids {sample_info_path}/update_ids.txt \\\n",
    "--make-bed \\\n",
    "--out {bed_dir}/indiv_samples/coriell_merge'\n",
    "    f.write(f\"{plink_merge_cmd}\")\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "e08141ba-563d-493a-b629-e444a753dfd8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19403938\n"
     ]
    }
   ],
   "source": [
    "!swarm -f {swarm_scripts_dir}/merge.swarm -g 64 -t 32 --time=10:00:00 --logdir swarm --gres=lscratch:20 --partition=norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d088b64a-86dc-4faf-a29f-dbe7597aa8a6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---- COMMAND EXECUTED: ---------------------------------------------------------\n",
      "(   plink --merge-list /data/CARD/PD/GP2/raw_genotypes/coriell/bed/merge_bed.list --update-ids /data/CARD/PD/GP2/raw_genotypes/coriell/sample_info/update_ids.txt --make-bed --out /data/CARD/PD/GP2/raw_genotypes/coriell/bed/indiv_samples/coriell_merge )\n",
      "--------------------------------------------------------------------------------\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /data/CARD/PD/GP2/raw_genotypes/coriell/bed/indiv_samples/coriell_merge.log.\n",
      "Options in effect:\n",
      "  --make-bed\n",
      "  --merge-list /data/CARD/PD/GP2/raw_genotypes/coriell/bed/merge_bed.list\n",
      "  --out /data/CARD/PD/GP2/raw_genotypes/coriell/bed/indiv_samples/coriell_merge\n",
      "  --update-ids /data/CARD/PD/GP2/raw_genotypes/coriell/sample_info/update_ids.txt\n",
      "\n",
      "257652 MB RAM detected; reserving 128826 MB for main workspace.\n",
      "Performing single-pass merge (4315 people, 2004347 variants).\n",
      "Merged fileset written to                     \n",
      "/data/CARD/PD/GP2/raw_genotypes/coriell/bed/indiv_samples/coriell_merge-merge.bed\n",
      "+\n",
      "/data/CARD/PD/GP2/raw_genotypes/coriell/bed/indiv_samples/coriell_merge-merge.bim\n",
      "+\n",
      "/data/CARD/PD/GP2/raw_genotypes/coriell/bed/indiv_samples/coriell_merge-merge.fam\n",
      ".\n",
      "2004347 variants loaded from .bim file.\n",
      "4315 people (2392 males, 1749 females, 174 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "/data/CARD/PD/GP2/raw_genotypes/coriell/bed/indiv_samples/coriell_merge.nosex .\n",
      "--update-ids: 4027 people updated, 1 ID not present.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 4315 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.985904.\n",
      "2004347 variants and 4315 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/data/CARD/PD/GP2/raw_genotypes/coriell/bed/indiv_samples/coriell_merge.bed +\n",
      "/data/CARD/PD/GP2/raw_genotypes/coriell/bed/indiv_samples/coriell_merge.bim +\n",
      "/data/CARD/PD/GP2/raw_genotypes/coriell/bed/indiv_samples/coriell_merge.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n"
     ]
    }
   ],
   "source": [
    "!cat swarm/swarm_19403938_0.o"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "52a39c0c-e013-4c59-b28e-203c0fb052c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /data/CARD/PD/GP2/raw_genotypes/coriell/bed/coriell.log.\n",
      "Options in effect:\n",
      "  --bfile /data/CARD/PD/GP2/raw_genotypes/coriell/bed/indiv_samples/coriell_merge\n",
      "  --make-bed\n",
      "  --out /data/CARD/PD/GP2/raw_genotypes/coriell/bed/coriell\n",
      "  --pheno /data/CARD/PD/GP2/raw_genotypes/coriell/sample_info/update_pheno.txt\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "2004347 variants loaded from .bim file.\n",
      "4315 people (2392 males, 1749 females, 174 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "/data/CARD/PD/GP2/raw_genotypes/coriell/bed/coriell.nosex .\n",
      "4027 phenotype values present after --pheno.\n",
      "Warning: Ignoring phenotypes of missing-sex samples.  If you don't want those\n",
      "phenotypes to be ignored, use the --allow-no-sex flag.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 4315 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Warning: 81699 het. haploid genotypes present (see\n",
      "/data/CARD/PD/GP2/raw_genotypes/coriell/bed/coriell.hh ); many commands treat\n",
      "these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Total genotyping rate is 0.985904.\n",
      "2004347 variants and 4315 people pass filters and QC.\n",
      "Among remaining phenotypes, 2777 are cases and 1250 are controls.  (288\n",
      "phenotypes are missing.)\n",
      "--make-bed to /data/CARD/PD/GP2/raw_genotypes/coriell/bed/coriell.bed +\n",
      "/data/CARD/PD/GP2/raw_genotypes/coriell/bed/coriell.bim +\n",
      "/data/CARD/PD/GP2/raw_genotypes/coriell/bed/coriell.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n"
     ]
    }
   ],
   "source": [
    "!plink --bfile {bed_dir}/indiv_samples/coriell_merge --pheno {sample_info_path}/update_pheno.txt --make-bed --out {bed_dir}/coriell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f80a04ba-da1e-45b3-b50b-3d7a56a7ecc7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "python3 ../run_qc_pipeline.py --geno /data/CARD/PD/GP2/raw_genotypes/coriell/bed/coriell --ref /data/LNG/vitaled2/1kgenomes/1kg_ashkj_ref_panel_gp2_pruned --ref_labels /data/LNG/vitaled2/1kgenomes/ref_panel_ancestry.txt --out /data/CARD/PD/GP2/genotypes/coriell/clean/coriell\n"
     ]
    }
   ],
   "source": [
    "# run QC pipeline\n",
    "geno_path = f'{basedir}/coriell/bed/coriell'\n",
    "out_dir = f'{out_genotypes}/coriell/clean'\n",
    "out_path = f'{out_dir}/coriell'\n",
    "os.makedirs(f'{out_dir}', exist_ok=True)\n",
    "ref_dir_path = '/data/LNG/vitaled2/1kgenomes'\n",
    "ref_panel = f'{ref_dir_path}/1kg_ashkj_ref_panel_gp2_pruned'\n",
    "ref_labels = f'{ref_dir_path}/ref_panel_ancestry.txt'\n",
    "\n",
    "\n",
    "with open(f'{swarm_scripts_dir}/run_qc_pipeline.swarm','w') as f:\n",
    "    run_pipeline = f'python3 ../run_qc_pipeline.py --geno {geno_path} --ref {ref_panel} --ref_labels {ref_labels} --out {out_path}'\n",
    "    f.write(f'{run_pipeline}\\n')\n",
    "f.close()\n",
    "!cat {swarm_scripts_dir}/run_qc_pipeline.swarm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c70c4cda-34b5-4ffd-9cdb-903843280848",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19417202\n"
     ]
    }
   ],
   "source": [
    "!swarm -f {swarm_scripts_dir}/run_qc_pipeline.swarm -g 64 -t 32 --time=10:00:00 --logdir swarm --gres=lscratch:20 --partition=norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54301a1e-acf1-4474-b320-f4bfbc3ea9d6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Executing: plink --bfile coriell_EAS --freq --out EAS\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to EAS.log.\n",
      "Options in effect:\n",
      "  --bfile coriell_EAS\n",
      "  --freq\n",
      "  --out EAS\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "1956936 variants loaded from .bim file.\n",
      "40 people (25 males, 15 females) loaded from .fam.\n",
      "37 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 40 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.99809.\n",
      "--freq: Allele frequencies (founders only) written to EAS.frq .\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Executing: perl HRC-1000G-check-bim.pl -b coriell_EAS.bim -f EAS.frq -r PASS.Variantsbravo-dbsnp-all.tab -h\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "         Script to check plink .bim files against HRC/1000G for\n",
      "        strand, id names, positions, alleles, ref/alt assignment\n",
      "                         William Rayner 2015\n",
      "                        wrayner@well.ox.ac.uk\n",
      "\n",
      "                             Version 4.2.5\n",
      "\n",
      "\n",
      "Options Set:\n",
      "Reference Panel:             HRC\n",
      "Bim filename:                coriell_EAS.bim\n",
      "Reference filename:          PASS.Variantsbravo-dbsnp-all.tab\n",
      "Allele frequencies filename: EAS.frq\n",
      "Allele frequency threshold:  0.2\n",
      "\n",
      "\n",
      "Reading PASS.Variantsbravo-dbsnp-all.tab\n",
      " 100000 200000 300000 400000 500000 600000 700000 800000 900000 1000000 1100000 1200000 1300000 1400000 1500000 1600000 1700000 1800000 1900000 2000000 2100000 2200000 2300000 2400000 2500000 2600000 2700000 2800000 2900000 3000000 3100000 3200000 3300000 3400000 3500000 3600000 3700000 3800000 3900000 4000000 4100000 4200000 4300000 4400000 4500000 4600000 4700000 4800000 4900000 5000000 5100000 5200000 5300000 5400000 5500000 5600000 5700000 5800000 5900000 6000000 6100000 6200000 6300000 6400000 6500000 6600000 6700000 6800000 6900000 7000000 7100000 7200000 7300000 7400000 7500000 7600000 7700000 7800000 7900000 8000000 8100000 8200000 8300000 8400000 8500000 8600000 8700000 8800000 8900000 9000000 9100000 9200000 9300000 9400000 9500000 9600000 9700000 9800000 9900000 10000000 10100000 10200000 10300000 10400000 10500000 10600000 10700000 10800000 10900000 11000000 11100000 11200000 11300000 11400000 11500000 11600000 11700000 11800000 11900000 12000000 12100000 12200000 12300000 12400000 12500000 12600000 12700000 12800000 12900000 13000000 13100000 13200000 13300000 13400000 13500000 13600000 13700000 13800000 13900000 14000000 14100000 14200000 14300000 14400000 14500000 14600000 14700000 14800000 14900000 15000000 15100000 15200000 15300000 15400000 15500000 15600000 15700000 15800000 15900000 16000000 16100000 16200000 16300000 16400000 16500000 16600000 16700000 16800000 16900000 17000000 17100000 17200000 17300000 17400000 17500000 17600000 17700000 17800000 17900000 18000000 18100000 18200000 18300000 18400000 18500000 18600000 18700000 18800000 18900000 19000000 19100000 19200000 19300000 19400000 19500000 19600000 19700000 19800000 19900000 20000000 20100000 20200000 20300000 20400000 20500000 20600000 20700000 20800000 20900000 21000000 21100000 21200000 21300000 21400000 21500000 21600000 21700000 21800000 21900000 22000000 22100000 22200000 22300000 22400000 22500000 22600000 22700000 22800000 22900000 23000000 23100000 23200000 23300000 23400000 23500000 23600000 23700000 23800000 23900000 24000000 24100000 24200000 24300000 24400000 24500000 24600000 24700000 24800000 24900000 25000000 25100000 25200000 25300000 25400000 25500000 25600000 25700000 25800000 25900000 26000000 26100000 26200000 26300000 26400000 26500000 26600000 26700000 26800000 26900000 27000000 27100000 27200000 27300000 27400000 27500000 27600000 27700000 27800000 27900000 28000000 28100000 28200000 28300000 28400000 28500000 28600000 28700000 28800000 28900000 29000000 29100000 29200000 29300000 29400000 29500000 29600000 29700000 29800000 29900000 30000000 30100000 30200000 30300000 30400000 30500000 30600000 30700000 30800000 30900000 31000000 31100000 31200000 31300000 31400000 31500000 31600000 31700000 31800000 31900000 32000000 32100000 32200000 32300000 32400000 32500000 32600000 32700000 32800000 32900000 33000000 33100000 33200000 33300000 33400000 33500000 33600000 33700000 33800000 33900000 34000000 34100000 34200000 34300000 34400000 34500000 34600000 34700000 34800000 34900000 35000000 35100000 35200000 35300000 35400000 35500000 35600000 35700000 35800000 35900000 36000000 36100000 36200000 36300000 36400000 36500000 36600000 36700000 36800000 36900000 37000000 37100000 37200000 37300000 37400000 37500000 37600000 37700000 37800000 37900000 38000000 38100000 38200000 38300000 38400000 38500000 38600000 38700000 38800000 38900000 39000000 39100000 39200000 39300000 39400000 39500000 39600000 39700000 39800000 39900000 40000000 40100000 40200000 40300000 40400000 40500000 40600000 40700000 40800000 40900000 41000000 41100000 41200000 41300000 41400000 41500000 41600000 41700000 41800000 41900000 42000000 42100000 42200000 42300000 42400000 42500000 42600000 42700000 42800000 42900000 43000000 43100000 43200000 43300000 43400000 43500000 43600000 43700000 43800000 43900000 44000000 44100000 44200000 44300000 44400000 44500000 44600000 44700000 44800000 44900000 45000000 45100000 45200000 45300000 45400000 45500000 45600000 45700000 45800000 45900000 46000000 46100000 46200000 46300000 46400000 46500000 46600000 46700000 46800000 46900000 47000000 47100000 47200000 47300000 47400000 47500000 47600000 47700000 47800000 47900000 48000000 48100000 48200000 48300000 48400000 48500000 48600000 48700000 48800000 48900000 49000000 49100000 49200000 49300000 49400000 49500000 49600000 49700000 49800000 49900000 50000000 50100000 50200000 50300000 50400000 50500000 50600000 50700000 50800000 50900000 51000000 51100000 51200000 51300000 51400000 51500000 51600000 51700000 51800000 51900000 52000000 52100000 52200000 52300000 52400000 52500000 52600000 52700000 52800000 52900000 53000000 53100000 53200000 53300000 53400000 53500000 53600000 53700000 53800000 53900000 54000000 54100000 54200000 54300000 54400000 54500000 54600000 54700000 54800000 54900000 55000000 55100000 55200000 55300000 55400000 55500000 55600000 55700000 55800000 55900000 56000000 56100000 56200000 56300000 56400000 56500000 56600000 56700000 56800000 56900000 57000000 57100000 57200000 57300000 57400000 57500000 57600000 57700000 57800000 57900000 58000000 58100000 58200000 58300000 58400000 58500000 58600000 58700000 58800000 58900000 59000000 59100000 59200000 59300000 59400000 59500000 59600000 59700000 59800000 59900000 60000000 60100000 60200000 60300000 60400000 60500000 60600000 60700000 60800000 60900000 61000000 61100000 61200000 61300000 61400000 61500000 61600000 61700000 61800000 61900000 62000000 62100000 62200000 62300000 62400000 62500000 62600000 62700000 62800000 62900000 63000000 63100000 63200000 63300000 63400000 63500000 63600000 63700000 63800000 63900000 64000000 64100000 64200000 64300000 64400000 64500000 64600000 64700000 64800000 64900000 65000000 65100000 65200000 65300000 65400000 65500000 65600000 65700000 65800000 65900000 66000000 66100000 66200000 66300000 66400000 66500000 66600000 66700000 66800000 66900000 67000000 67100000 67200000 67300000 67400000 67500000 67600000 67700000 67800000 67900000 68000000 68100000 68200000 68300000 68400000 68500000 68600000 68700000 68800000 68900000 69000000 69100000 69200000 69300000 69400000 69500000 69600000 69700000 69800000 69900000 70000000 70100000 70200000 70300000 70400000 70500000 70600000 70700000 70800000 70900000 71000000 71100000 71200000 71300000 71400000 71500000 71600000 71700000 71800000 71900000 72000000 72100000 72200000 72300000 72400000 72500000 72600000 72700000 72800000 72900000 73000000 73100000 73200000 73300000 73400000 73500000 73600000 73700000 73800000 73900000 74000000 74100000 74200000 74300000 74400000 74500000 74600000 74700000 74800000 74900000 75000000 75100000 75200000 75300000 75400000 75500000 75600000 75700000 75800000 75900000 76000000 76100000 76200000 76300000 76400000 76500000 76600000 76700000 76800000 76900000 77000000 77100000 77200000 77300000 77400000 77500000 77600000 77700000 77800000 77900000 78000000 78100000 78200000 78300000 78400000 78500000 78600000 78700000 78800000 78900000 79000000 79100000 79200000 79300000 79400000 79500000 79600000 79700000 79800000 79900000 80000000 80100000 80200000 80300000 80400000 80500000 80600000 80700000 80800000 80900000 81000000 81100000 81200000 81300000 81400000 81500000 81600000 81700000 81800000 81900000 82000000 82100000 82200000 82300000 82400000 82500000 82600000 82700000 82800000 82900000 83000000 83100000 83200000 83300000 83400000 83500000 83600000 83700000 83800000 83900000 84000000 84100000 84200000 84300000 84400000 84500000 84600000 84700000 84800000 84900000 85000000 85100000 85200000 85300000 85400000 85500000 85600000 85700000 85800000 85900000 86000000 86100000 86200000 86300000 86400000 86500000 86600000 86700000 86800000 86900000 87000000 87100000 87200000 87300000 87400000 87500000 87600000 87700000 87800000 87900000 88000000 88100000 88200000 88300000 88400000 88500000 88600000 88700000 88800000 88900000 89000000 89100000 89200000 89300000 89400000 89500000 89600000 89700000 89800000 89900000 90000000 90100000 90200000 90300000 90400000 90500000 90600000 90700000 90800000 90900000 91000000 91100000 91200000 91300000 91400000 91500000 91600000 91700000 91800000 91900000 92000000 92100000 92200000 92300000 92400000 92500000 92600000 92700000 92800000 92900000 93000000 93100000 93200000 93300000 93400000 93500000 93600000 93700000 93800000 93900000 94000000 94100000 94200000 94300000 94400000 94500000 94600000 94700000 94800000 94900000 95000000 95100000 95200000 95300000 95400000 95500000 95600000 95700000 95800000 95900000 96000000 96100000 96200000 96300000 96400000 96500000 96600000 96700000 96800000 96900000 97000000 97100000 97200000 97300000 97400000 97500000 97600000 97700000 97800000 97900000 98000000 98100000 98200000 98300000 98400000 98500000 98600000 98700000 98800000 98900000 99000000 99100000 99200000 99300000 99400000 99500000 99600000 99700000 99800000 99900000 100000000 100100000 100200000 100300000 100400000 100500000 100600000 100700000 100800000 100900000 101000000 101100000 101200000 101300000 101400000 101500000 101600000 101700000 101800000 101900000 102000000 102100000 102200000 102300000 102400000 102500000 102600000 102700000 102800000 102900000 103000000 103100000 103200000 103300000 103400000 103500000 103600000 103700000 103800000 103900000 104000000 104100000 104200000 104300000 104400000 104500000 104600000 104700000 104800000 104900000 105000000 105100000 105200000 105300000 105400000 105500000 105600000 105700000 105800000 105900000 106000000 106100000 106200000 106300000 106400000 106500000 106600000 106700000 106800000 106900000 107000000 107100000 107200000 107300000 107400000 107500000 107600000 107700000 107800000 107900000 108000000 108100000 108200000 108300000 108400000 108500000 108600000 108700000 108800000 108900000 109000000 109100000 109200000 109300000 109400000 109500000 109600000 109700000 109800000 109900000 110000000 110100000 110200000 110300000 110400000 110500000 110600000 110700000 110800000 110900000 111000000 111100000 111200000 111300000 111400000 111500000 111600000 111700000 111800000 111900000 112000000 112100000 112200000 112300000 112400000 112500000 112600000 112700000 112800000 112900000 113000000 113100000 113200000 113300000 113400000 113500000 113600000 113700000 113800000 113900000 114000000 114100000 114200000 114300000 114400000 114500000 114600000 114700000 114800000 114900000 115000000 115100000 115200000 115300000 115400000 115500000 115600000 115700000 115800000 115900000 116000000 116100000 116200000 116300000 116400000 116500000 116600000 116700000 116800000 116900000 117000000 117100000 117200000 117300000 117400000 117500000 117600000 117700000 117800000 117900000 118000000 118100000 118200000 118300000 118400000 118500000 118600000 118700000 118800000 118900000 119000000 119100000 119200000 119300000 119400000 119500000 119600000 119700000 119800000 119900000 120000000 120100000 120200000 120300000 120400000 120500000 120600000 120700000 120800000 120900000 121000000 121100000 121200000 121300000 121400000 121500000 121600000 121700000 121800000 121900000 122000000 122100000 122200000 122300000 122400000 122500000 122600000 122700000 122800000 122900000 123000000 123100000 123200000 123300000 123400000 123500000 123600000 123700000 123800000 123900000 124000000 124100000 124200000 124300000 124400000 124500000 124600000 124700000 124800000 124900000 125000000 125100000 125200000 125300000 125400000 125500000 125600000 125700000 125800000 125900000 126000000 126100000 126200000 126300000 126400000 126500000 126600000 126700000 126800000 126900000 127000000 127100000 127200000 127300000 127400000 127500000 127600000 127700000 127800000 127900000 128000000 128100000 128200000 128300000 128400000 128500000 128600000 128700000 128800000 128900000 129000000 129100000 129200000 129300000 129400000 129500000 129600000 129700000 129800000 129900000 130000000 130100000 130200000 130300000 130400000 130500000 130600000 130700000 130800000 130900000 131000000 131100000 131200000 131300000 131400000 131500000 131600000 131700000 131800000 131900000 132000000 132100000 132200000 132300000 132400000 132500000 132600000 132700000 132800000 132900000 133000000 133100000 133200000 133300000 133400000 133500000 133600000 133700000 133800000 133900000 134000000 134100000 134200000 134300000 134400000 134500000 134600000 134700000 134800000 134900000 135000000 135100000 135200000 135300000 135400000 135500000 135600000 135700000 135800000 135900000 136000000 136100000 136200000 136300000 136400000 136500000 136600000 136700000 136800000 136900000 137000000 137100000 137200000 137300000 137400000 137500000 137600000 137700000 137800000 137900000 138000000 138100000 138200000 138300000 138400000 138500000 138600000 138700000 138800000 138900000 139000000 139100000 139200000 139300000 139400000 139500000 139600000 139700000 139800000 139900000 140000000 140100000 140200000 140300000 140400000 140500000 140600000 140700000 140800000 140900000 141000000 141100000 141200000 141300000 141400000 141500000 141600000 141700000 141800000 141900000 142000000 142100000 142200000 142300000 142400000 142500000 142600000 142700000 142800000 142900000 143000000 143100000 143200000 143300000 143400000 143500000 143600000 143700000 143800000 143900000 144000000 144100000 144200000 144300000 144400000 144500000 144600000 144700000 144800000 144900000 145000000 145100000 145200000 145300000 145400000 145500000 145600000 145700000 145800000 145900000 146000000 146100000 146200000 146300000 146400000 146500000 146600000 146700000 146800000 146900000 147000000 147100000 147200000 147300000 147400000 147500000 147600000 147700000 147800000 147900000 148000000 148100000 148200000 148300000 148400000 148500000 148600000 148700000 148800000 148900000 149000000 149100000 149200000 149300000 149400000 149500000 149600000 149700000 149800000 149900000 150000000 150100000 150200000 150300000 150400000 150500000 150600000 150700000 150800000 150900000 151000000 151100000 151200000 151300000 151400000 151500000 151600000 151700000 151800000 151900000 152000000 152100000 152200000 152300000 152400000 152500000 152600000 152700000 152800000 152900000 153000000 153100000 153200000 153300000 153400000 153500000 153600000 153700000 153800000 153900000 154000000 154100000 154200000 154300000 154400000 154500000 154600000 154700000 154800000 154900000 155000000 155100000 155200000 155300000 155400000 155500000 155600000 155700000 155800000 155900000 156000000 156100000 156200000 156300000 156400000 156500000 156600000 156700000 156800000 156900000 157000000 157100000 157200000 157300000 157400000 157500000 157600000 157700000 157800000 157900000 158000000 158100000 158200000 158300000 158400000 158500000 158600000 158700000 158800000 158900000 159000000 159100000 159200000 159300000 159400000 159500000 159600000 159700000 159800000 159900000 160000000 160100000 160200000 160300000 160400000 160500000 160600000 160700000 160800000 160900000 161000000 161100000 161200000 161300000 161400000 161500000 161600000 161700000 161800000 161900000 162000000 162100000 162200000 162300000 162400000 162500000 162600000 162700000 162800000 162900000 163000000 163100000 163200000 163300000 163400000 163500000 163600000 163700000 163800000 163900000 164000000 164100000 164200000 164300000 164400000 164500000 164600000 164700000 164800000 164900000 165000000 165100000 165200000 165300000 165400000 165500000 165600000 165700000 165800000 165900000 166000000 166100000 166200000 166300000 166400000 166500000 166600000 166700000 166800000 166900000 167000000 167100000 167200000 167300000 167400000 167500000 167600000 167700000 167800000 167900000 168000000 168100000 168200000 168300000 168400000 168500000 168600000 168700000 168800000 168900000 169000000 169100000 169200000 169300000 169400000 169500000 169600000 169700000 169800000 169900000 170000000 Done\n",
      "Matching to HRC\n",
      "\n",
      "Position Matches\n",
      " ID matches HRC 744797\n",
      " ID Doesn't match HRC 708484\n",
      " Total Position Matches 1453281\n",
      "ID Match\n",
      " Different position to HRC 118\n",
      "No Match to HRC 393126\n",
      "Skipped (X, XY, Y, MT) 13137\n",
      "Total in bim file 1956936\n",
      "Total processed 1859662\n",
      "\n",
      "Indels (ignored in r1) 45081\n",
      "\n",
      "SNPs not changed 156378\n",
      "SNPs to change ref alt 1138848\n",
      "Strand ok 756885\n",
      "Total Strand ok 1295226\n",
      "\n",
      "Strand to change 621788\n",
      "Total checked 1453399\n",
      "Total checked Strand 1378673\n",
      "Total removed for allele Frequency diff > 0.2 214966\n",
      "Palindromic SNPs with Freq > 0.4 6799\n",
      "\n",
      "\n",
      "Non Matching alleles 67927\n",
      "ID and allele mismatching 62438; where HRC is . 36831\n",
      "Duplicates removed 52193\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Executing: sh Run-plink.sh\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to TEMP1.log.\n",
      "Options in effect:\n",
      "  --bfile coriell_EAS\n",
      "  --exclude Exclude-coriell_EAS-HRC.txt\n",
      "  --make-bed\n",
      "  --out TEMP1\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "1956936 variants loaded from .bim file.\n",
      "40 people (25 males, 15 females) loaded from .fam.\n",
      "37 phenotype values loaded from .fam.\n",
      "--exclude: 1163707 variants remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 40 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998225.\n",
      "1163707 variants and 40 people pass filters and QC.\n",
      "Among remaining phenotypes, 28 are cases and 9 are controls.  (3 phenotypes are\n",
      "missing.)\n",
      "--make-bed to TEMP1.bed + TEMP1.bim + TEMP1.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to TEMP2.log.\n",
      "Options in effect:\n",
      "  --bfile TEMP1\n",
      "  --make-bed\n",
      "  --out TEMP2\n",
      "  --update-chr\n",
      "  --update-map Chromosome-coriell_EAS-HRC.txt\n",
      "\n",
      "Note: --update-map [filename] + parameter-free --update-chr deprecated.  Use\n",
      "--update-chr [filename] instead.\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "1163707 variants loaded from .bim file.\n",
      "40 people (25 males, 15 females) loaded from .fam.\n",
      "37 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 40 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998225.\n",
      "1163707 variants and 40 people pass filters and QC.\n",
      "Among remaining phenotypes, 28 are cases and 9 are controls.  (3 phenotypes are\n",
      "missing.)\n",
      "--update-chr: 16 values updated, 102 variant IDs not present.\n",
      "--make-bed to TEMP2.bed + TEMP2.bim + TEMP2.fam ... 1010111112121313141415151616171718181919202021212222232324242525262627272828292930303131323233333434353536363737383839394040414142424343444445454646474748484949505051515252535354545555565657575858595960606161626263636464656566666767686869697070717172727373747475757676777778787979808081818282838384848585868687878888898990909191929293939494959596969797989899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to TEMP3.log.\n",
      "Options in effect:\n",
      "  --bfile TEMP2\n",
      "  --make-bed\n",
      "  --out TEMP3\n",
      "  --update-map Position-coriell_EAS-HRC.txt\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "1163707 variants loaded from .bim file.\n",
      "40 people (25 males, 15 females) loaded from .fam.\n",
      "37 phenotype values loaded from .fam.\n",
      "--update-map: 16 values updated, 102 variant IDs not present.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 40 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998225.\n",
      "1163707 variants and 40 people pass filters and QC.\n",
      "Among remaining phenotypes, 28 are cases and 9 are controls.  (3 phenotypes are\n",
      "missing.)\n",
      "--make-bed to TEMP3.bed + TEMP3.bim + TEMP3.fam ... 1010111112121313141415151616171718181919202021212222232324242525262627272828292930303131323233333434353536363737383839394040414142424343444445454646474748484949505051515252535354545555565657575858595960606161626263636464656566666767686869697070717172727373747475757676777778787979808081818282838384848585868687878888898990909191929293939494959596969797989899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to TEMP4.log.\n",
      "Options in effect:\n",
      "  --bfile TEMP3\n",
      "  --flip Strand-Flip-coriell_EAS-HRC.txt\n",
      "  --make-bed\n",
      "  --out TEMP4\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "1163707 variants loaded from .bim file.\n",
      "40 people (25 males, 15 females) loaded from .fam.\n",
      "37 phenotype values loaded from .fam.\n",
      "--flip: 552323 SNPs flipped, 69465 SNP IDs not present.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 40 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998225.\n",
      "1163707 variants and 40 people pass filters and QC.\n",
      "Among remaining phenotypes, 28 are cases and 9 are controls.  (3 phenotypes are\n",
      "missing.)\n",
      "--make-bed to TEMP4.bed + TEMP4.bim + TEMP4.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_EAS-updated.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_EAS-HRC.txt\n",
      "  --bfile TEMP4\n",
      "  --make-bed\n",
      "  --out coriell_EAS-updated\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "1163707 variants loaded from .bim file.\n",
      "40 people (25 males, 15 females) loaded from .fam.\n",
      "37 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 40 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998225.\n",
      "--a1-allele: 1163707 assignments made.\n",
      "1163707 variants and 40 people pass filters and QC.\n",
      "Among remaining phenotypes, 28 are cases and 9 are controls.  (3 phenotypes are\n",
      "missing.)\n",
      "--make-bed to coriell_EAS-updated.bed + coriell_EAS-updated.bim +\n",
      "coriell_EAS-updated.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_EAS-updated-chr1.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_EAS-HRC.txt\n",
      "  --bfile coriell_EAS-updated\n",
      "  --chr 1\n",
      "  --make-bed\n",
      "  --out coriell_EAS-updated-chr1\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "94429 out of 1163707 variants loaded from .bim file.\n",
      "40 people (25 males, 15 females) loaded from .fam.\n",
      "37 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 40 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998224.\n",
      "--a1-allele: 94429 assignments made.\n",
      "94429 variants and 40 people pass filters and QC.\n",
      "Among remaining phenotypes, 28 are cases and 9 are controls.  (3 phenotypes are\n",
      "missing.)\n",
      "--make-bed to coriell_EAS-updated-chr1.bed + coriell_EAS-updated-chr1.bim +\n",
      "coriell_EAS-updated-chr1.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_EAS-updated-chr2.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_EAS-HRC.txt\n",
      "  --bfile coriell_EAS-updated\n",
      "  --chr 2\n",
      "  --make-bed\n",
      "  --out coriell_EAS-updated-chr2\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "98627 out of 1163707 variants loaded from .bim file.\n",
      "40 people (25 males, 15 females) loaded from .fam.\n",
      "37 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 40 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998513.\n",
      "--a1-allele: 98627 assignments made.\n",
      "98627 variants and 40 people pass filters and QC.\n",
      "Among remaining phenotypes, 28 are cases and 9 are controls.  (3 phenotypes are\n",
      "missing.)\n",
      "--make-bed to coriell_EAS-updated-chr2.bed + coriell_EAS-updated-chr2.bim +\n",
      "coriell_EAS-updated-chr2.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_EAS-updated-chr3.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_EAS-HRC.txt\n",
      "  --bfile coriell_EAS-updated\n",
      "  --chr 3\n",
      "  --make-bed\n",
      "  --out coriell_EAS-updated-chr3\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "83016 out of 1163707 variants loaded from .bim file.\n",
      "40 people (25 males, 15 females) loaded from .fam.\n",
      "37 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 40 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998435.\n",
      "--a1-allele: 83016 assignments made.\n",
      "83016 variants and 40 people pass filters and QC.\n",
      "Among remaining phenotypes, 28 are cases and 9 are controls.  (3 phenotypes are\n",
      "missing.)\n",
      "--make-bed to coriell_EAS-updated-chr3.bed + coriell_EAS-updated-chr3.bim +\n",
      "coriell_EAS-updated-chr3.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_EAS-updated-chr4.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_EAS-HRC.txt\n",
      "  --bfile coriell_EAS-updated\n",
      "  --chr 4\n",
      "  --make-bed\n",
      "  --out coriell_EAS-updated-chr4\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "74844 out of 1163707 variants loaded from .bim file.\n",
      "40 people (25 males, 15 females) loaded from .fam.\n",
      "37 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 40 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.99841.\n",
      "--a1-allele: 74844 assignments made.\n",
      "74844 variants and 40 people pass filters and QC.\n",
      "Among remaining phenotypes, 28 are cases and 9 are controls.  (3 phenotypes are\n",
      "missing.)\n",
      "--make-bed to coriell_EAS-updated-chr4.bed + coriell_EAS-updated-chr4.bim +\n",
      "coriell_EAS-updated-chr4.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_EAS-updated-chr5.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_EAS-HRC.txt\n",
      "  --bfile coriell_EAS-updated\n",
      "  --chr 5\n",
      "  --make-bed\n",
      "  --out coriell_EAS-updated-chr5\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "68282 out of 1163707 variants loaded from .bim file.\n",
      "40 people (25 males, 15 females) loaded from .fam.\n",
      "37 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 40 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998385.\n",
      "--a1-allele: 68282 assignments made.\n",
      "68282 variants and 40 people pass filters and QC.\n",
      "Among remaining phenotypes, 28 are cases and 9 are controls.  (3 phenotypes are\n",
      "missing.)\n",
      "--make-bed to coriell_EAS-updated-chr5.bed + coriell_EAS-updated-chr5.bim +\n",
      "coriell_EAS-updated-chr5.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_EAS-updated-chr6.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_EAS-HRC.txt\n",
      "  --bfile coriell_EAS-updated\n",
      "  --chr 6\n",
      "  --make-bed\n",
      "  --out coriell_EAS-updated-chr6\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "78356 out of 1163707 variants loaded from .bim file.\n",
      "40 people (25 males, 15 females) loaded from .fam.\n",
      "37 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 40 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998124.\n",
      "--a1-allele: 78356 assignments made.\n",
      "78356 variants and 40 people pass filters and QC.\n",
      "Among remaining phenotypes, 28 are cases and 9 are controls.  (3 phenotypes are\n",
      "missing.)\n",
      "--make-bed to coriell_EAS-updated-chr6.bed + coriell_EAS-updated-chr6.bim +\n",
      "coriell_EAS-updated-chr6.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_EAS-updated-chr7.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_EAS-HRC.txt\n",
      "  --bfile coriell_EAS-updated\n",
      "  --chr 7\n",
      "  --make-bed\n",
      "  --out coriell_EAS-updated-chr7\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "64069 out of 1163707 variants loaded from .bim file.\n",
      "40 people (25 males, 15 females) loaded from .fam.\n",
      "37 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 40 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998081.\n",
      "--a1-allele: 64069 assignments made.\n",
      "64069 variants and 40 people pass filters and QC.\n",
      "Among remaining phenotypes, 28 are cases and 9 are controls.  (3 phenotypes are\n",
      "missing.)\n",
      "--make-bed to coriell_EAS-updated-chr7.bed + coriell_EAS-updated-chr7.bim +\n",
      "coriell_EAS-updated-chr7.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_EAS-updated-chr8.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_EAS-HRC.txt\n",
      "  --bfile coriell_EAS-updated\n",
      "  --chr 8\n",
      "  --make-bed\n",
      "  --out coriell_EAS-updated-chr8\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "58482 out of 1163707 variants loaded from .bim file.\n",
      "40 people (25 males, 15 females) loaded from .fam.\n",
      "37 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 40 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.99825.\n",
      "--a1-allele: 58482 assignments made.\n",
      "58482 variants and 40 people pass filters and QC.\n",
      "Among remaining phenotypes, 28 are cases and 9 are controls.  (3 phenotypes are\n",
      "missing.)\n",
      "--make-bed to coriell_EAS-updated-chr8.bed + coriell_EAS-updated-chr8.bim +\n",
      "coriell_EAS-updated-chr8.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_EAS-updated-chr9.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_EAS-HRC.txt\n",
      "  --bfile coriell_EAS-updated\n",
      "  --chr 9\n",
      "  --make-bed\n",
      "  --out coriell_EAS-updated-chr9\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "49822 out of 1163707 variants loaded from .bim file.\n",
      "40 people (25 males, 15 females) loaded from .fam.\n",
      "37 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 40 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998166.\n",
      "--a1-allele: 49822 assignments made.\n",
      "49822 variants and 40 people pass filters and QC.\n",
      "Among remaining phenotypes, 28 are cases and 9 are controls.  (3 phenotypes are\n",
      "missing.)\n",
      "--make-bed to coriell_EAS-updated-chr9.bed + coriell_EAS-updated-chr9.bim +\n",
      "coriell_EAS-updated-chr9.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_EAS-updated-chr10.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_EAS-HRC.txt\n",
      "  --bfile coriell_EAS-updated\n",
      "  --chr 10\n",
      "  --make-bed\n",
      "  --out coriell_EAS-updated-chr10\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "54878 out of 1163707 variants loaded from .bim file.\n",
      "40 people (25 males, 15 females) loaded from .fam.\n",
      "37 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 40 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998276.\n",
      "--a1-allele: 54878 assignments made.\n",
      "54878 variants and 40 people pass filters and QC.\n",
      "Among remaining phenotypes, 28 are cases and 9 are controls.  (3 phenotypes are\n",
      "missing.)\n",
      "--make-bed to coriell_EAS-updated-chr10.bed + coriell_EAS-updated-chr10.bim +\n",
      "coriell_EAS-updated-chr10.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_EAS-updated-chr11.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_EAS-HRC.txt\n",
      "  --bfile coriell_EAS-updated\n",
      "  --chr 11\n",
      "  --make-bed\n",
      "  --out coriell_EAS-updated-chr11\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "60930 out of 1163707 variants loaded from .bim file.\n",
      "40 people (25 males, 15 females) loaded from .fam.\n",
      "37 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 40 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998199.\n",
      "--a1-allele: 60930 assignments made.\n",
      "60930 variants and 40 people pass filters and QC.\n",
      "Among remaining phenotypes, 28 are cases and 9 are controls.  (3 phenotypes are\n",
      "missing.)\n",
      "--make-bed to coriell_EAS-updated-chr11.bed + coriell_EAS-updated-chr11.bim +\n",
      "coriell_EAS-updated-chr11.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_EAS-updated-chr12.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_EAS-HRC.txt\n",
      "  --bfile coriell_EAS-updated\n",
      "  --chr 12\n",
      "  --make-bed\n",
      "  --out coriell_EAS-updated-chr12\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "55150 out of 1163707 variants loaded from .bim file.\n",
      "40 people (25 males, 15 females) loaded from .fam.\n",
      "37 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 40 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.99807.\n",
      "--a1-allele: 55150 assignments made.\n",
      "55150 variants and 40 people pass filters and QC.\n",
      "Among remaining phenotypes, 28 are cases and 9 are controls.  (3 phenotypes are\n",
      "missing.)\n",
      "--make-bed to coriell_EAS-updated-chr12.bed + coriell_EAS-updated-chr12.bim +\n",
      "coriell_EAS-updated-chr12.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_EAS-updated-chr13.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_EAS-HRC.txt\n",
      "  --bfile coriell_EAS-updated\n",
      "  --chr 13\n",
      "  --make-bed\n",
      "  --out coriell_EAS-updated-chr13\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "38139 out of 1163707 variants loaded from .bim file.\n",
      "40 people (25 males, 15 females) loaded from .fam.\n",
      "37 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 40 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998589.\n",
      "--a1-allele: 38139 assignments made.\n",
      "38139 variants and 40 people pass filters and QC.\n",
      "Among remaining phenotypes, 28 are cases and 9 are controls.  (3 phenotypes are\n",
      "missing.)\n",
      "--make-bed to coriell_EAS-updated-chr13.bed + coriell_EAS-updated-chr13.bim +\n",
      "coriell_EAS-updated-chr13.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_EAS-updated-chr14.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_EAS-HRC.txt\n",
      "  --bfile coriell_EAS-updated\n",
      "  --chr 14\n",
      "  --make-bed\n",
      "  --out coriell_EAS-updated-chr14\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "36870 out of 1163707 variants loaded from .bim file.\n",
      "40 people (25 males, 15 females) loaded from .fam.\n",
      "37 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 40 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998403.\n",
      "--a1-allele: 36870 assignments made.\n",
      "36870 variants and 40 people pass filters and QC.\n",
      "Among remaining phenotypes, 28 are cases and 9 are controls.  (3 phenotypes are\n",
      "missing.)\n",
      "--make-bed to coriell_EAS-updated-chr14.bed + coriell_EAS-updated-chr14.bim +\n",
      "coriell_EAS-updated-chr14.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_EAS-updated-chr15.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_EAS-HRC.txt\n",
      "  --bfile coriell_EAS-updated\n",
      "  --chr 15\n",
      "  --make-bed\n",
      "  --out coriell_EAS-updated-chr15\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "35415 out of 1163707 variants loaded from .bim file.\n",
      "40 people (25 males, 15 females) loaded from .fam.\n",
      "37 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 40 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998244.\n",
      "--a1-allele: 35415 assignments made.\n",
      "35415 variants and 40 people pass filters and QC.\n",
      "Among remaining phenotypes, 28 are cases and 9 are controls.  (3 phenotypes are\n",
      "missing.)\n",
      "--make-bed to coriell_EAS-updated-chr15.bed + coriell_EAS-updated-chr15.bim +\n",
      "coriell_EAS-updated-chr15.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_EAS-updated-chr16.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_EAS-HRC.txt\n",
      "  --bfile coriell_EAS-updated\n",
      "  --chr 16\n",
      "  --make-bed\n",
      "  --out coriell_EAS-updated-chr16\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "40354 out of 1163707 variants loaded from .bim file.\n",
      "40 people (25 males, 15 females) loaded from .fam.\n",
      "37 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 40 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.997845.\n",
      "--a1-allele: 40354 assignments made.\n",
      "40354 variants and 40 people pass filters and QC.\n",
      "Among remaining phenotypes, 28 are cases and 9 are controls.  (3 phenotypes are\n",
      "missing.)\n",
      "--make-bed to coriell_EAS-updated-chr16.bed + coriell_EAS-updated-chr16.bim +\n",
      "coriell_EAS-updated-chr16.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_EAS-updated-chr17.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_EAS-HRC.txt\n",
      "  --bfile coriell_EAS-updated\n",
      "  --chr 17\n",
      "  --make-bed\n",
      "  --out coriell_EAS-updated-chr17\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "39838 out of 1163707 variants loaded from .bim file.\n",
      "40 people (25 males, 15 females) loaded from .fam.\n",
      "37 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 40 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.997815.\n",
      "--a1-allele: 39838 assignments made.\n",
      "39838 variants and 40 people pass filters and QC.\n",
      "Among remaining phenotypes, 28 are cases and 9 are controls.  (3 phenotypes are\n",
      "missing.)\n",
      "--make-bed to coriell_EAS-updated-chr17.bed + coriell_EAS-updated-chr17.bim +\n",
      "coriell_EAS-updated-chr17.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_EAS-updated-chr18.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_EAS-HRC.txt\n",
      "  --bfile coriell_EAS-updated\n",
      "  --chr 18\n",
      "  --make-bed\n",
      "  --out coriell_EAS-updated-chr18\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "31737 out of 1163707 variants loaded from .bim file.\n",
      "40 people (25 males, 15 females) loaded from .fam.\n",
      "37 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 40 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998601.\n",
      "--a1-allele: 31737 assignments made.\n",
      "31737 variants and 40 people pass filters and QC.\n",
      "Among remaining phenotypes, 28 are cases and 9 are controls.  (3 phenotypes are\n",
      "missing.)\n",
      "--make-bed to coriell_EAS-updated-chr18.bed + coriell_EAS-updated-chr18.bim +\n",
      "coriell_EAS-updated-chr18.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_EAS-updated-chr19.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_EAS-HRC.txt\n",
      "  --bfile coriell_EAS-updated\n",
      "  --chr 19\n",
      "  --make-bed\n",
      "  --out coriell_EAS-updated-chr19\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "33197 out of 1163707 variants loaded from .bim file.\n",
      "40 people (25 males, 15 females) loaded from .fam.\n",
      "37 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 40 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.997348.\n",
      "--a1-allele: 33197 assignments made.\n",
      "33197 variants and 40 people pass filters and QC.\n",
      "Among remaining phenotypes, 28 are cases and 9 are controls.  (3 phenotypes are\n",
      "missing.)\n",
      "--make-bed to coriell_EAS-updated-chr19.bed + coriell_EAS-updated-chr19.bim +\n",
      "coriell_EAS-updated-chr19.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_EAS-updated-chr20.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_EAS-HRC.txt\n",
      "  --bfile coriell_EAS-updated\n",
      "  --chr 20\n",
      "  --make-bed\n",
      "  --out coriell_EAS-updated-chr20\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "27956 out of 1163707 variants loaded from .bim file.\n",
      "40 people (25 males, 15 females) loaded from .fam.\n",
      "37 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 40 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998193.\n",
      "--a1-allele: 27956 assignments made.\n",
      "27956 variants and 40 people pass filters and QC.\n",
      "Among remaining phenotypes, 28 are cases and 9 are controls.  (3 phenotypes are\n",
      "missing.)\n",
      "--make-bed to coriell_EAS-updated-chr20.bed + coriell_EAS-updated-chr20.bim +\n",
      "coriell_EAS-updated-chr20.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_EAS-updated-chr21.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_EAS-HRC.txt\n",
      "  --bfile coriell_EAS-updated\n",
      "  --chr 21\n",
      "  --make-bed\n",
      "  --out coriell_EAS-updated-chr21\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "16306 out of 1163707 variants loaded from .bim file.\n",
      "40 people (25 males, 15 females) loaded from .fam.\n",
      "37 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 40 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998235.\n",
      "--a1-allele: 16306 assignments made.\n",
      "16306 variants and 40 people pass filters and QC.\n",
      "Among remaining phenotypes, 28 are cases and 9 are controls.  (3 phenotypes are\n",
      "missing.)\n",
      "--make-bed to coriell_EAS-updated-chr21.bed + coriell_EAS-updated-chr21.bim +\n",
      "coriell_EAS-updated-chr21.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_EAS-updated-chr22.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_EAS-HRC.txt\n",
      "  --bfile coriell_EAS-updated\n",
      "  --chr 22\n",
      "  --make-bed\n",
      "  --out coriell_EAS-updated-chr22\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "17553 out of 1163707 variants loaded from .bim file.\n",
      "40 people (25 males, 15 females) loaded from .fam.\n",
      "37 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 40 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.99753.\n",
      "--a1-allele: 17553 assignments made.\n",
      "17553 variants and 40 people pass filters and QC.\n",
      "Among remaining phenotypes, 28 are cases and 9 are controls.  (3 phenotypes are\n",
      "missing.)\n",
      "--make-bed to coriell_EAS-updated-chr22.bed + coriell_EAS-updated-chr22.bim +\n",
      "coriell_EAS-updated-chr22.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_EAS-updated-chr23.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_EAS-HRC.txt\n",
      "  --bfile coriell_EAS-updated\n",
      "  --chr 23\n",
      "  --make-bed\n",
      "  --out coriell_EAS-updated-chr23\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "5457 out of 1163707 variants loaded from .bim file.\n",
      "40 people (25 males, 15 females) loaded from .fam.\n",
      "37 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 40 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.997489.\n",
      "--a1-allele: 5457 assignments made.\n",
      "5457 variants and 40 people pass filters and QC.\n",
      "Among remaining phenotypes, 28 are cases and 9 are controls.  (3 phenotypes are\n",
      "missing.)\n",
      "--make-bed to coriell_EAS-updated-chr23.bed + coriell_EAS-updated-chr23.bim +\n",
      "coriell_EAS-updated-chr23.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Executing: plink --bfile coriell_EAS-updated-chr1 --recode vcf --chr 1 --out EAS_chr1\n",
      "Executing: plink --bfile coriell_EAS-updated-chr2 --recode vcf --chr 2 --out EAS_chr2\n",
      "Executing: plink --bfile coriell_EAS-updated-chr3 --recode vcf --chr 3 --out EAS_chr3\n",
      "Executing: plink --bfile coriell_EAS-updated-chr4 --recode vcf --chr 4 --out EAS_chr4\n",
      "Executing: plink --bfile coriell_EAS-updated-chr5 --recode vcf --chr 5 --out EAS_chr5\n",
      "Executing: plink --bfile coriell_EAS-updated-chr6 --recode vcf --chr 6 --out EAS_chr6\n",
      "Executing: plink --bfile coriell_EAS-updated-chr7 --recode vcf --chr 7 --out EAS_chr7\n",
      "Executing: plink --bfile coriell_EAS-updated-chr8 --recode vcf --chr 8 --out EAS_chr8\n",
      "Executing: plink --bfile coriell_EAS-updated-chr9 --recode vcf --chr 9 --out EAS_chr9\n",
      "Executing: plink --bfile coriell_EAS-updated-chr10 --recode vcf --chr 10 --out EAS_chr10\n",
      "Executing: plink --bfile coriell_EAS-updated-chr11 --recode vcf --chr 11 --out EAS_chr11\n",
      "Executing: plink --bfile coriell_EAS-updated-chr12 --recode vcf --chr 12 --out EAS_chr12\n",
      "Executing: plink --bfile coriell_EAS-updated-chr13 --recode vcf --chr 13 --out EAS_chr13\n",
      "Executing: plink --bfile coriell_EAS-updated-chr14 --recode vcf --chr 14 --out EAS_chr14\n",
      "Executing: plink --bfile coriell_EAS-updated-chr15 --recode vcf --chr 15 --out EAS_chr15\n",
      "Executing: plink --bfile coriell_EAS-updated-chr16 --recode vcf --chr 16 --out EAS_chr16\n",
      "Executing: plink --bfile coriell_EAS-updated-chr17 --recode vcf --chr 17 --out EAS_chr17\n",
      "Executing: plink --bfile coriell_EAS-updated-chr18 --recode vcf --chr 18 --out EAS_chr18\n",
      "Executing: plink --bfile coriell_EAS-updated-chr19 --recode vcf --chr 19 --out EAS_chr19\n",
      "Executing: plink --bfile coriell_EAS-updated-chr20 --recode vcf --chr 20 --out EAS_chr20\n",
      "Executing: plink --bfile coriell_EAS-updated-chr21 --recode vcf --chr 21 --out EAS_chr21\n",
      "Executing: plink --bfile coriell_EAS-updated-chr22 --recode vcf --chr 22 --out EAS_chr22\n",
      "Executing: plink --bfile coriell_EAS-updated-chr23 --recode vcf --chr 23 --out EAS_chr23\n",
      "Executing: plink --bfile coriell_AJ --freq --out AJ\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to AJ.log.\n",
      "Options in effect:\n",
      "  --bfile coriell_AJ\n",
      "  --freq\n",
      "  --out AJ\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "1942242 variants loaded from .bim file.\n",
      "517 people (323 males, 194 females) loaded from .fam.\n",
      "432 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 517 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998167.\n",
      "--freq: Allele frequencies (founders only) written to AJ.frq .\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Executing: perl HRC-1000G-check-bim.pl -b coriell_AJ.bim -f AJ.frq -r PASS.Variantsbravo-dbsnp-all.tab -h\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "         Script to check plink .bim files against HRC/1000G for\n",
      "        strand, id names, positions, alleles, ref/alt assignment\n",
      "                         William Rayner 2015\n",
      "                        wrayner@well.ox.ac.uk\n",
      "\n",
      "                             Version 4.2.5\n",
      "\n",
      "\n",
      "Options Set:\n",
      "Reference Panel:             HRC\n",
      "Bim filename:                coriell_AJ.bim\n",
      "Reference filename:          PASS.Variantsbravo-dbsnp-all.tab\n",
      "Allele frequencies filename: AJ.frq\n",
      "Allele frequency threshold:  0.2\n",
      "\n",
      "\n",
      "Reading PASS.Variantsbravo-dbsnp-all.tab\n",
      " 100000 200000 300000 400000 500000 600000 700000 800000 900000 1000000 1100000 1200000 1300000 1400000 1500000 1600000 1700000 1800000 1900000 2000000 2100000 2200000 2300000 2400000 2500000 2600000 2700000 2800000 2900000 3000000 3100000 3200000 3300000 3400000 3500000 3600000 3700000 3800000 3900000 4000000 4100000 4200000 4300000 4400000 4500000 4600000 4700000 4800000 4900000 5000000 5100000 5200000 5300000 5400000 5500000 5600000 5700000 5800000 5900000 6000000 6100000 6200000 6300000 6400000 6500000 6600000 6700000 6800000 6900000 7000000 7100000 7200000 7300000 7400000 7500000 7600000 7700000 7800000 7900000 8000000 8100000 8200000 8300000 8400000 8500000 8600000 8700000 8800000 8900000 9000000 9100000 9200000 9300000 9400000 9500000 9600000 9700000 9800000 9900000 10000000 10100000 10200000 10300000 10400000 10500000 10600000 10700000 10800000 10900000 11000000 11100000 11200000 11300000 11400000 11500000 11600000 11700000 11800000 11900000 12000000 12100000 12200000 12300000 12400000 12500000 12600000 12700000 12800000 12900000 13000000 13100000 13200000 13300000 13400000 13500000 13600000 13700000 13800000 13900000 14000000 14100000 14200000 14300000 14400000 14500000 14600000 14700000 14800000 14900000 15000000 15100000 15200000 15300000 15400000 15500000 15600000 15700000 15800000 15900000 16000000 16100000 16200000 16300000 16400000 16500000 16600000 16700000 16800000 16900000 17000000 17100000 17200000 17300000 17400000 17500000 17600000 17700000 17800000 17900000 18000000 18100000 18200000 18300000 18400000 18500000 18600000 18700000 18800000 18900000 19000000 19100000 19200000 19300000 19400000 19500000 19600000 19700000 19800000 19900000 20000000 20100000 20200000 20300000 20400000 20500000 20600000 20700000 20800000 20900000 21000000 21100000 21200000 21300000 21400000 21500000 21600000 21700000 21800000 21900000 22000000 22100000 22200000 22300000 22400000 22500000 22600000 22700000 22800000 22900000 23000000 23100000 23200000 23300000 23400000 23500000 23600000 23700000 23800000 23900000 24000000 24100000 24200000 24300000 24400000 24500000 24600000 24700000 24800000 24900000 25000000 25100000 25200000 25300000 25400000 25500000 25600000 25700000 25800000 25900000 26000000 26100000 26200000 26300000 26400000 26500000 26600000 26700000 26800000 26900000 27000000 27100000 27200000 27300000 27400000 27500000 27600000 27700000 27800000 27900000 28000000 28100000 28200000 28300000 28400000 28500000 28600000 28700000 28800000 28900000 29000000 29100000 29200000 29300000 29400000 29500000 29600000 29700000 29800000 29900000 30000000 30100000 30200000 30300000 30400000 30500000 30600000 30700000 30800000 30900000 31000000 31100000 31200000 31300000 31400000 31500000 31600000 31700000 31800000 31900000 32000000 32100000 32200000 32300000 32400000 32500000 32600000 32700000 32800000 32900000 33000000 33100000 33200000 33300000 33400000 33500000 33600000 33700000 33800000 33900000 34000000 34100000 34200000 34300000 34400000 34500000 34600000 34700000 34800000 34900000 35000000 35100000 35200000 35300000 35400000 35500000 35600000 35700000 35800000 35900000 36000000 36100000 36200000 36300000 36400000 36500000 36600000 36700000 36800000 36900000 37000000 37100000 37200000 37300000 37400000 37500000 37600000 37700000 37800000 37900000 38000000 38100000 38200000 38300000 38400000 38500000 38600000 38700000 38800000 38900000 39000000 39100000 39200000 39300000 39400000 39500000 39600000 39700000 39800000 39900000 40000000 40100000 40200000 40300000 40400000 40500000 40600000 40700000 40800000 40900000 41000000 41100000 41200000 41300000 41400000 41500000 41600000 41700000 41800000 41900000 42000000 42100000 42200000 42300000 42400000 42500000 42600000 42700000 42800000 42900000 43000000 43100000 43200000 43300000 43400000 43500000 43600000 43700000 43800000 43900000 44000000 44100000 44200000 44300000 44400000 44500000 44600000 44700000 44800000 44900000 45000000 45100000 45200000 45300000 45400000 45500000 45600000 45700000 45800000 45900000 46000000 46100000 46200000 46300000 46400000 46500000 46600000 46700000 46800000 46900000 47000000 47100000 47200000 47300000 47400000 47500000 47600000 47700000 47800000 47900000 48000000 48100000 48200000 48300000 48400000 48500000 48600000 48700000 48800000 48900000 49000000 49100000 49200000 49300000 49400000 49500000 49600000 49700000 49800000 49900000 50000000 50100000 50200000 50300000 50400000 50500000 50600000 50700000 50800000 50900000 51000000 51100000 51200000 51300000 51400000 51500000 51600000 51700000 51800000 51900000 52000000 52100000 52200000 52300000 52400000 52500000 52600000 52700000 52800000 52900000 53000000 53100000 53200000 53300000 53400000 53500000 53600000 53700000 53800000 53900000 54000000 54100000 54200000 54300000 54400000 54500000 54600000 54700000 54800000 54900000 55000000 55100000 55200000 55300000 55400000 55500000 55600000 55700000 55800000 55900000 56000000 56100000 56200000 56300000 56400000 56500000 56600000 56700000 56800000 56900000 57000000 57100000 57200000 57300000 57400000 57500000 57600000 57700000 57800000 57900000 58000000 58100000 58200000 58300000 58400000 58500000 58600000 58700000 58800000 58900000 59000000 59100000 59200000 59300000 59400000 59500000 59600000 59700000 59800000 59900000 60000000 60100000 60200000 60300000 60400000 60500000 60600000 60700000 60800000 60900000 61000000 61100000 61200000 61300000 61400000 61500000 61600000 61700000 61800000 61900000 62000000 62100000 62200000 62300000 62400000 62500000 62600000 62700000 62800000 62900000 63000000 63100000 63200000 63300000 63400000 63500000 63600000 63700000 63800000 63900000 64000000 64100000 64200000 64300000 64400000 64500000 64600000 64700000 64800000 64900000 65000000 65100000 65200000 65300000 65400000 65500000 65600000 65700000 65800000 65900000 66000000 66100000 66200000 66300000 66400000 66500000 66600000 66700000 66800000 66900000 67000000 67100000 67200000 67300000 67400000 67500000 67600000 67700000 67800000 67900000 68000000 68100000 68200000 68300000 68400000 68500000 68600000 68700000 68800000 68900000 69000000 69100000 69200000 69300000 69400000 69500000 69600000 69700000 69800000 69900000 70000000 70100000 70200000 70300000 70400000 70500000 70600000 70700000 70800000 70900000 71000000 71100000 71200000 71300000 71400000 71500000 71600000 71700000 71800000 71900000 72000000 72100000 72200000 72300000 72400000 72500000 72600000 72700000 72800000 72900000 73000000 73100000 73200000 73300000 73400000 73500000 73600000 73700000 73800000 73900000 74000000 74100000 74200000 74300000 74400000 74500000 74600000 74700000 74800000 74900000 75000000 75100000 75200000 75300000 75400000 75500000 75600000 75700000 75800000 75900000 76000000 76100000 76200000 76300000 76400000 76500000 76600000 76700000 76800000 76900000 77000000 77100000 77200000 77300000 77400000 77500000 77600000 77700000 77800000 77900000 78000000 78100000 78200000 78300000 78400000 78500000 78600000 78700000 78800000 78900000 79000000 79100000 79200000 79300000 79400000 79500000 79600000 79700000 79800000 79900000 80000000 80100000 80200000 80300000 80400000 80500000 80600000 80700000 80800000 80900000 81000000 81100000 81200000 81300000 81400000 81500000 81600000 81700000 81800000 81900000 82000000 82100000 82200000 82300000 82400000 82500000 82600000 82700000 82800000 82900000 83000000 83100000 83200000 83300000 83400000 83500000 83600000 83700000 83800000 83900000 84000000 84100000 84200000 84300000 84400000 84500000 84600000 84700000 84800000 84900000 85000000 85100000 85200000 85300000 85400000 85500000 85600000 85700000 85800000 85900000 86000000 86100000 86200000 86300000 86400000 86500000 86600000 86700000 86800000 86900000 87000000 87100000 87200000 87300000 87400000 87500000 87600000 87700000 87800000 87900000 88000000 88100000 88200000 88300000 88400000 88500000 88600000 88700000 88800000 88900000 89000000 89100000 89200000 89300000 89400000 89500000 89600000 89700000 89800000 89900000 90000000 90100000 90200000 90300000 90400000 90500000 90600000 90700000 90800000 90900000 91000000 91100000 91200000 91300000 91400000 91500000 91600000 91700000 91800000 91900000 92000000 92100000 92200000 92300000 92400000 92500000 92600000 92700000 92800000 92900000 93000000 93100000 93200000 93300000 93400000 93500000 93600000 93700000 93800000 93900000 94000000 94100000 94200000 94300000 94400000 94500000 94600000 94700000 94800000 94900000 95000000 95100000 95200000 95300000 95400000 95500000 95600000 95700000 95800000 95900000 96000000 96100000 96200000 96300000 96400000 96500000 96600000 96700000 96800000 96900000 97000000 97100000 97200000 97300000 97400000 97500000 97600000 97700000 97800000 97900000 98000000 98100000 98200000 98300000 98400000 98500000 98600000 98700000 98800000 98900000 99000000 99100000 99200000 99300000 99400000 99500000 99600000 99700000 99800000 99900000 100000000 100100000 100200000 100300000 100400000 100500000 100600000 100700000 100800000 100900000 101000000 101100000 101200000 101300000 101400000 101500000 101600000 101700000 101800000 101900000 102000000 102100000 102200000 102300000 102400000 102500000 102600000 102700000 102800000 102900000 103000000 103100000 103200000 103300000 103400000 103500000 103600000 103700000 103800000 103900000 104000000 104100000 104200000 104300000 104400000 104500000 104600000 104700000 104800000 104900000 105000000 105100000 105200000 105300000 105400000 105500000 105600000 105700000 105800000 105900000 106000000 106100000 106200000 106300000 106400000 106500000 106600000 106700000 106800000 106900000 107000000 107100000 107200000 107300000 107400000 107500000 107600000 107700000 107800000 107900000 108000000 108100000 108200000 108300000 108400000 108500000 108600000 108700000 108800000 108900000 109000000 109100000 109200000 109300000 109400000 109500000 109600000 109700000 109800000 109900000 110000000 110100000 110200000 110300000 110400000 110500000 110600000 110700000 110800000 110900000 111000000 111100000 111200000 111300000 111400000 111500000 111600000 111700000 111800000 111900000 112000000 112100000 112200000 112300000 112400000 112500000 112600000 112700000 112800000 112900000 113000000 113100000 113200000 113300000 113400000 113500000 113600000 113700000 113800000 113900000 114000000 114100000 114200000 114300000 114400000 114500000 114600000 114700000 114800000 114900000 115000000 115100000 115200000 115300000 115400000 115500000 115600000 115700000 115800000 115900000 116000000 116100000 116200000 116300000 116400000 116500000 116600000 116700000 116800000 116900000 117000000 117100000 117200000 117300000 117400000 117500000 117600000 117700000 117800000 117900000 118000000 118100000 118200000 118300000 118400000 118500000 118600000 118700000 118800000 118900000 119000000 119100000 119200000 119300000 119400000 119500000 119600000 119700000 119800000 119900000 120000000 120100000 120200000 120300000 120400000 120500000 120600000 120700000 120800000 120900000 121000000 121100000 121200000 121300000 121400000 121500000 121600000 121700000 121800000 121900000 122000000 122100000 122200000 122300000 122400000 122500000 122600000 122700000 122800000 122900000 123000000 123100000 123200000 123300000 123400000 123500000 123600000 123700000 123800000 123900000 124000000 124100000 124200000 124300000 124400000 124500000 124600000 124700000 124800000 124900000 125000000 125100000 125200000 125300000 125400000 125500000 125600000 125700000 125800000 125900000 126000000 126100000 126200000 126300000 126400000 126500000 126600000 126700000 126800000 126900000 127000000 127100000 127200000 127300000 127400000 127500000 127600000 127700000 127800000 127900000 128000000 128100000 128200000 128300000 128400000 128500000 128600000 128700000 128800000 128900000 129000000 129100000 129200000 129300000 129400000 129500000 129600000 129700000 129800000 129900000 130000000 130100000 130200000 130300000 130400000 130500000 130600000 130700000 130800000 130900000 131000000 131100000 131200000 131300000 131400000 131500000 131600000 131700000 131800000 131900000 132000000 132100000 132200000 132300000 132400000 132500000 132600000 132700000 132800000 132900000 133000000 133100000 133200000 133300000 133400000 133500000 133600000 133700000 133800000 133900000 134000000 134100000 134200000 134300000 134400000 134500000 134600000 134700000 134800000 134900000 135000000 135100000 135200000 135300000 135400000 135500000 135600000 135700000 135800000 135900000 136000000 136100000 136200000 136300000 136400000 136500000 136600000 136700000 136800000 136900000 137000000 137100000 137200000 137300000 137400000 137500000 137600000 137700000 137800000 137900000 138000000 138100000 138200000 138300000 138400000 138500000 138600000 138700000 138800000 138900000 139000000 139100000 139200000 139300000 139400000 139500000 139600000 139700000 139800000 139900000 140000000 140100000 140200000 140300000 140400000 140500000 140600000 140700000 140800000 140900000 141000000 141100000 141200000 141300000 141400000 141500000 141600000 141700000 141800000 141900000 142000000 142100000 142200000 142300000 142400000 142500000 142600000 142700000 142800000 142900000 143000000 143100000 143200000 143300000 143400000 143500000 143600000 143700000 143800000 143900000 144000000 144100000 144200000 144300000 144400000 144500000 144600000 144700000 144800000 144900000 145000000 145100000 145200000 145300000 145400000 145500000 145600000 145700000 145800000 145900000 146000000 146100000 146200000 146300000 146400000 146500000 146600000 146700000 146800000 146900000 147000000 147100000 147200000 147300000 147400000 147500000 147600000 147700000 147800000 147900000 148000000 148100000 148200000 148300000 148400000 148500000 148600000 148700000 148800000 148900000 149000000 149100000 149200000 149300000 149400000 149500000 149600000 149700000 149800000 149900000 150000000 150100000 150200000 150300000 150400000 150500000 150600000 150700000 150800000 150900000 151000000 151100000 151200000 151300000 151400000 151500000 151600000 151700000 151800000 151900000 152000000 152100000 152200000 152300000 152400000 152500000 152600000 152700000 152800000 152900000 153000000 153100000 153200000 153300000 153400000 153500000 153600000 153700000 153800000 153900000 154000000 154100000 154200000 154300000 154400000 154500000 154600000 154700000 154800000 154900000 155000000 155100000 155200000 155300000 155400000 155500000 155600000 155700000 155800000 155900000 156000000 156100000 156200000 156300000 156400000 156500000 156600000 156700000 156800000 156900000 157000000 157100000 157200000 157300000 157400000 157500000 157600000 157700000 157800000 157900000 158000000 158100000 158200000 158300000 158400000 158500000 158600000 158700000 158800000 158900000 159000000 159100000 159200000 159300000 159400000 159500000 159600000 159700000 159800000 159900000 160000000 160100000 160200000 160300000 160400000 160500000 160600000 160700000 160800000 160900000 161000000 161100000 161200000 161300000 161400000 161500000 161600000 161700000 161800000 161900000 162000000 162100000 162200000 162300000 162400000 162500000 162600000 162700000 162800000 162900000 163000000 163100000 163200000 163300000 163400000 163500000 163600000 163700000 163800000 163900000 164000000 164100000 164200000 164300000 164400000 164500000 164600000 164700000 164800000 164900000 165000000 165100000 165200000 165300000 165400000 165500000 165600000 165700000 165800000 165900000 166000000 166100000 166200000 166300000 166400000 166500000 166600000 166700000 166800000 166900000 167000000 167100000 167200000 167300000 167400000 167500000 167600000 167700000 167800000 167900000 168000000 168100000 168200000 168300000 168400000 168500000 168600000 168700000 168800000 168900000 169000000 169100000 169200000 169300000 169400000 169500000 169600000 169700000 169800000 169900000 170000000 Done\n",
      "Matching to HRC\n",
      "\n",
      "Position Matches\n",
      " ID matches HRC 740460\n",
      " ID Doesn't match HRC 702805\n",
      " Total Position Matches 1443265\n",
      "ID Match\n",
      " Different position to HRC 118\n",
      "No Match to HRC 389765\n",
      "Skipped (X, XY, Y, MT) 12969\n",
      "Total in bim file 1942242\n",
      "Total processed 1846117\n",
      "\n",
      "Indels (ignored in r1) 44878\n",
      "\n",
      "SNPs not changed 148767\n",
      "SNPs to change ref alt 1144326\n",
      "Strand ok 751936\n",
      "Total Strand ok 1293093\n",
      "\n",
      "Strand to change 617242\n",
      "Total checked 1443383\n",
      "Total checked Strand 1369178\n",
      "Total removed for allele Frequency diff > 0.2 91176\n",
      "Palindromic SNPs with Freq > 0.4 6703\n",
      "\n",
      "\n",
      "Non Matching alleles 67502\n",
      "ID and allele mismatching 62035; where HRC is . 36543\n",
      "Duplicates removed 51247\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Executing: sh Run-plink.sh\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to TEMP1.log.\n",
      "Options in effect:\n",
      "  --bfile coriell_AJ\n",
      "  --exclude Exclude-coriell_AJ-HRC.txt\n",
      "  --make-bed\n",
      "  --out TEMP1\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "1942242 variants loaded from .bim file.\n",
      "517 people (323 males, 194 females) loaded from .fam.\n",
      "432 phenotype values loaded from .fam.\n",
      "--exclude: 1278002 variants remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 517 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998309.\n",
      "1278002 variants and 517 people pass filters and QC.\n",
      "Among remaining phenotypes, 325 are cases and 107 are controls.  (85 phenotypes\n",
      "are missing.)\n",
      "--make-bed to TEMP1.bed + TEMP1.bim + TEMP1.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to TEMP2.log.\n",
      "Options in effect:\n",
      "  --bfile TEMP1\n",
      "  --make-bed\n",
      "  --out TEMP2\n",
      "  --update-chr\n",
      "  --update-map Chromosome-coriell_AJ-HRC.txt\n",
      "\n",
      "Note: --update-map [filename] + parameter-free --update-chr deprecated.  Use\n",
      "--update-chr [filename] instead.\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "1278002 variants loaded from .bim file.\n",
      "517 people (323 males, 194 females) loaded from .fam.\n",
      "432 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 517 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998309.\n",
      "1278002 variants and 517 people pass filters and QC.\n",
      "Among remaining phenotypes, 325 are cases and 107 are controls.  (85 phenotypes\n",
      "are missing.)\n",
      "--update-chr: 16 values updated, 102 variant IDs not present.\n",
      "--make-bed to TEMP2.bed + TEMP2.bim + TEMP2.fam ... 10101111121213131414151516161717181819192020212122222323242425252626272728282929303031313232333334343535363637373838393940404141424243434444454546464747484849505051515252535354545555565657575858595960606161626263636464656566666767686869697070717172727373747475757676777778787979808081818282838384848585868687878888898990909191929293939494959596969797989899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to TEMP3.log.\n",
      "Options in effect:\n",
      "  --bfile TEMP2\n",
      "  --make-bed\n",
      "  --out TEMP3\n",
      "  --update-map Position-coriell_AJ-HRC.txt\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "1278002 variants loaded from .bim file.\n",
      "517 people (323 males, 194 females) loaded from .fam.\n",
      "432 phenotype values loaded from .fam.\n",
      "--update-map: 16 values updated, 102 variant IDs not present.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 517 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998309.\n",
      "1278002 variants and 517 people pass filters and QC.\n",
      "Among remaining phenotypes, 325 are cases and 107 are controls.  (85 phenotypes\n",
      "are missing.)\n",
      "--make-bed to TEMP3.bed + TEMP3.bim + TEMP3.fam ... 10101111121213131414151516161717181819192020212122222323242425252626272728282929303031313232333334343535363637373838393940404141424243434444454546464747484849505051515252535354545555565657575858595960606161626263636464656566666767686869697070717172727373747475757676777778787979808081818282838384848585868687878888898990909191929293939494959596969797989899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to TEMP4.log.\n",
      "Options in effect:\n",
      "  --bfile TEMP3\n",
      "  --flip Strand-Flip-coriell_AJ-HRC.txt\n",
      "  --make-bed\n",
      "  --out TEMP4\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "1278002 variants loaded from .bim file.\n",
      "517 people (323 males, 194 females) loaded from .fam.\n",
      "432 phenotype values loaded from .fam.\n",
      "--flip: 607798 SNPs flipped, 9444 SNP IDs not present.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 517 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998309.\n",
      "1278002 variants and 517 people pass filters and QC.\n",
      "Among remaining phenotypes, 325 are cases and 107 are controls.  (85 phenotypes\n",
      "are missing.)\n",
      "--make-bed to TEMP4.bed + TEMP4.bim + TEMP4.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_AJ-updated.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_AJ-HRC.txt\n",
      "  --bfile TEMP4\n",
      "  --make-bed\n",
      "  --out coriell_AJ-updated\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "1278002 variants loaded from .bim file.\n",
      "517 people (323 males, 194 females) loaded from .fam.\n",
      "432 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 517 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998309.\n",
      "--a1-allele: 1278002 assignments made.\n",
      "1278002 variants and 517 people pass filters and QC.\n",
      "Among remaining phenotypes, 325 are cases and 107 are controls.  (85 phenotypes\n",
      "are missing.)\n",
      "--make-bed to coriell_AJ-updated.bed + coriell_AJ-updated.bim +\n",
      "coriell_AJ-updated.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_AJ-updated-chr1.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_AJ-HRC.txt\n",
      "  --bfile coriell_AJ-updated\n",
      "  --chr 1\n",
      "  --make-bed\n",
      "  --out coriell_AJ-updated-chr1\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "103456 out of 1278002 variants loaded from .bim file.\n",
      "517 people (323 males, 194 females) loaded from .fam.\n",
      "432 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 517 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998229.\n",
      "--a1-allele: 103456 assignments made.\n",
      "103456 variants and 517 people pass filters and QC.\n",
      "Among remaining phenotypes, 325 are cases and 107 are controls.  (85 phenotypes\n",
      "are missing.)\n",
      "--make-bed to coriell_AJ-updated-chr1.bed + coriell_AJ-updated-chr1.bim +\n",
      "coriell_AJ-updated-chr1.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_AJ-updated-chr2.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_AJ-HRC.txt\n",
      "  --bfile coriell_AJ-updated\n",
      "  --chr 2\n",
      "  --make-bed\n",
      "  --out coriell_AJ-updated-chr2\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "109314 out of 1278002 variants loaded from .bim file.\n",
      "517 people (323 males, 194 females) loaded from .fam.\n",
      "432 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 517 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998621.\n",
      "--a1-allele: 109314 assignments made.\n",
      "109314 variants and 517 people pass filters and QC.\n",
      "Among remaining phenotypes, 325 are cases and 107 are controls.  (85 phenotypes\n",
      "are missing.)\n",
      "--make-bed to coriell_AJ-updated-chr2.bed + coriell_AJ-updated-chr2.bim +\n",
      "coriell_AJ-updated-chr2.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_AJ-updated-chr3.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_AJ-HRC.txt\n",
      "  --bfile coriell_AJ-updated\n",
      "  --chr 3\n",
      "  --make-bed\n",
      "  --out coriell_AJ-updated-chr3\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "91712 out of 1278002 variants loaded from .bim file.\n",
      "517 people (323 males, 194 females) loaded from .fam.\n",
      "432 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 517 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998574.\n",
      "--a1-allele: 91712 assignments made.\n",
      "91712 variants and 517 people pass filters and QC.\n",
      "Among remaining phenotypes, 325 are cases and 107 are controls.  (85 phenotypes\n",
      "are missing.)\n",
      "--make-bed to coriell_AJ-updated-chr3.bed + coriell_AJ-updated-chr3.bim +\n",
      "coriell_AJ-updated-chr3.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_AJ-updated-chr4.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_AJ-HRC.txt\n",
      "  --bfile coriell_AJ-updated\n",
      "  --chr 4\n",
      "  --make-bed\n",
      "  --out coriell_AJ-updated-chr4\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "82405 out of 1278002 variants loaded from .bim file.\n",
      "517 people (323 males, 194 females) loaded from .fam.\n",
      "432 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 517 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998489.\n",
      "--a1-allele: 82405 assignments made.\n",
      "82405 variants and 517 people pass filters and QC.\n",
      "Among remaining phenotypes, 325 are cases and 107 are controls.  (85 phenotypes\n",
      "are missing.)\n",
      "--make-bed to coriell_AJ-updated-chr4.bed + coriell_AJ-updated-chr4.bim +\n",
      "coriell_AJ-updated-chr4.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_AJ-updated-chr5.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_AJ-HRC.txt\n",
      "  --bfile coriell_AJ-updated\n",
      "  --chr 5\n",
      "  --make-bed\n",
      "  --out coriell_AJ-updated-chr5\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "75248 out of 1278002 variants loaded from .bim file.\n",
      "517 people (323 males, 194 females) loaded from .fam.\n",
      "432 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 517 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998544.\n",
      "--a1-allele: 75248 assignments made.\n",
      "75248 variants and 517 people pass filters and QC.\n",
      "Among remaining phenotypes, 325 are cases and 107 are controls.  (85 phenotypes\n",
      "are missing.)\n",
      "--make-bed to coriell_AJ-updated-chr5.bed + coriell_AJ-updated-chr5.bim +\n",
      "coriell_AJ-updated-chr5.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_AJ-updated-chr6.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_AJ-HRC.txt\n",
      "  --bfile coriell_AJ-updated\n",
      "  --chr 6\n",
      "  --make-bed\n",
      "  --out coriell_AJ-updated-chr6\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "85710 out of 1278002 variants loaded from .bim file.\n",
      "517 people (323 males, 194 females) loaded from .fam.\n",
      "432 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 517 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998228.\n",
      "--a1-allele: 85710 assignments made.\n",
      "85710 variants and 517 people pass filters and QC.\n",
      "Among remaining phenotypes, 325 are cases and 107 are controls.  (85 phenotypes\n",
      "are missing.)\n",
      "--make-bed to coriell_AJ-updated-chr6.bed + coriell_AJ-updated-chr6.bim +\n",
      "coriell_AJ-updated-chr6.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_AJ-updated-chr7.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_AJ-HRC.txt\n",
      "  --bfile coriell_AJ-updated\n",
      "  --chr 7\n",
      "  --make-bed\n",
      "  --out coriell_AJ-updated-chr7\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "70681 out of 1278002 variants loaded from .bim file.\n",
      "517 people (323 males, 194 females) loaded from .fam.\n",
      "432 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 517 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998164.\n",
      "--a1-allele: 70681 assignments made.\n",
      "70681 variants and 517 people pass filters and QC.\n",
      "Among remaining phenotypes, 325 are cases and 107 are controls.  (85 phenotypes\n",
      "are missing.)\n",
      "--make-bed to coriell_AJ-updated-chr7.bed + coriell_AJ-updated-chr7.bim +\n",
      "coriell_AJ-updated-chr7.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_AJ-updated-chr8.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_AJ-HRC.txt\n",
      "  --bfile coriell_AJ-updated\n",
      "  --chr 8\n",
      "  --make-bed\n",
      "  --out coriell_AJ-updated-chr8\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "64330 out of 1278002 variants loaded from .bim file.\n",
      "517 people (323 males, 194 females) loaded from .fam.\n",
      "432 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 517 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998309.\n",
      "--a1-allele: 64330 assignments made.\n",
      "64330 variants and 517 people pass filters and QC.\n",
      "Among remaining phenotypes, 325 are cases and 107 are controls.  (85 phenotypes\n",
      "are missing.)\n",
      "--make-bed to coriell_AJ-updated-chr8.bed + coriell_AJ-updated-chr8.bim +\n",
      "coriell_AJ-updated-chr8.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_AJ-updated-chr9.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_AJ-HRC.txt\n",
      "  --bfile coriell_AJ-updated\n",
      "  --chr 9\n",
      "  --make-bed\n",
      "  --out coriell_AJ-updated-chr9\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "54891 out of 1278002 variants loaded from .bim file.\n",
      "517 people (323 males, 194 females) loaded from .fam.\n",
      "432 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 517 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998187.\n",
      "--a1-allele: 54891 assignments made.\n",
      "54891 variants and 517 people pass filters and QC.\n",
      "Among remaining phenotypes, 325 are cases and 107 are controls.  (85 phenotypes\n",
      "are missing.)\n",
      "--make-bed to coriell_AJ-updated-chr9.bed + coriell_AJ-updated-chr9.bim +\n",
      "coriell_AJ-updated-chr9.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_AJ-updated-chr10.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_AJ-HRC.txt\n",
      "  --bfile coriell_AJ-updated\n",
      "  --chr 10\n",
      "  --make-bed\n",
      "  --out coriell_AJ-updated-chr10\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "60472 out of 1278002 variants loaded from .bim file.\n",
      "517 people (323 males, 194 females) loaded from .fam.\n",
      "432 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 517 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998415.\n",
      "--a1-allele: 60472 assignments made.\n",
      "60472 variants and 517 people pass filters and QC.\n",
      "Among remaining phenotypes, 325 are cases and 107 are controls.  (85 phenotypes\n",
      "are missing.)\n",
      "--make-bed to coriell_AJ-updated-chr10.bed + coriell_AJ-updated-chr10.bim +\n",
      "coriell_AJ-updated-chr10.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_AJ-updated-chr11.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_AJ-HRC.txt\n",
      "  --bfile coriell_AJ-updated\n",
      "  --chr 11\n",
      "  --make-bed\n",
      "  --out coriell_AJ-updated-chr11\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "66327 out of 1278002 variants loaded from .bim file.\n",
      "517 people (323 males, 194 females) loaded from .fam.\n",
      "432 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 517 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998326.\n",
      "--a1-allele: 66327 assignments made.\n",
      "66327 variants and 517 people pass filters and QC.\n",
      "Among remaining phenotypes, 325 are cases and 107 are controls.  (85 phenotypes\n",
      "are missing.)\n",
      "--make-bed to coriell_AJ-updated-chr11.bed + coriell_AJ-updated-chr11.bim +\n",
      "coriell_AJ-updated-chr11.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_AJ-updated-chr12.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_AJ-HRC.txt\n",
      "  --bfile coriell_AJ-updated\n",
      "  --chr 12\n",
      "  --make-bed\n",
      "  --out coriell_AJ-updated-chr12\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "60540 out of 1278002 variants loaded from .bim file.\n",
      "517 people (323 males, 194 females) loaded from .fam.\n",
      "432 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 517 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998426.\n",
      "--a1-allele: 60540 assignments made.\n",
      "60540 variants and 517 people pass filters and QC.\n",
      "Among remaining phenotypes, 325 are cases and 107 are controls.  (85 phenotypes\n",
      "are missing.)\n",
      "--make-bed to coriell_AJ-updated-chr12.bed + coriell_AJ-updated-chr12.bim +\n",
      "coriell_AJ-updated-chr12.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_AJ-updated-chr13.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_AJ-HRC.txt\n",
      "  --bfile coriell_AJ-updated\n",
      "  --chr 13\n",
      "  --make-bed\n",
      "  --out coriell_AJ-updated-chr13\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "42270 out of 1278002 variants loaded from .bim file.\n",
      "517 people (323 males, 194 females) loaded from .fam.\n",
      "432 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 517 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998667.\n",
      "--a1-allele: 42270 assignments made.\n",
      "42270 variants and 517 people pass filters and QC.\n",
      "Among remaining phenotypes, 325 are cases and 107 are controls.  (85 phenotypes\n",
      "are missing.)\n",
      "--make-bed to coriell_AJ-updated-chr13.bed + coriell_AJ-updated-chr13.bim +\n",
      "coriell_AJ-updated-chr13.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_AJ-updated-chr14.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_AJ-HRC.txt\n",
      "  --bfile coriell_AJ-updated\n",
      "  --chr 14\n",
      "  --make-bed\n",
      "  --out coriell_AJ-updated-chr14\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "40415 out of 1278002 variants loaded from .bim file.\n",
      "517 people (323 males, 194 females) loaded from .fam.\n",
      "432 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 517 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998429.\n",
      "--a1-allele: 40415 assignments made.\n",
      "40415 variants and 517 people pass filters and QC.\n",
      "Among remaining phenotypes, 325 are cases and 107 are controls.  (85 phenotypes\n",
      "are missing.)\n",
      "--make-bed to coriell_AJ-updated-chr14.bed + coriell_AJ-updated-chr14.bim +\n",
      "coriell_AJ-updated-chr14.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_AJ-updated-chr15.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_AJ-HRC.txt\n",
      "  --bfile coriell_AJ-updated\n",
      "  --chr 15\n",
      "  --make-bed\n",
      "  --out coriell_AJ-updated-chr15\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "39086 out of 1278002 variants loaded from .bim file.\n",
      "517 people (323 males, 194 females) loaded from .fam.\n",
      "432 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 517 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998421.\n",
      "--a1-allele: 39086 assignments made.\n",
      "39086 variants and 517 people pass filters and QC.\n",
      "Among remaining phenotypes, 325 are cases and 107 are controls.  (85 phenotypes\n",
      "are missing.)\n",
      "--make-bed to coriell_AJ-updated-chr15.bed + coriell_AJ-updated-chr15.bim +\n",
      "coriell_AJ-updated-chr15.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_AJ-updated-chr16.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_AJ-HRC.txt\n",
      "  --bfile coriell_AJ-updated\n",
      "  --chr 16\n",
      "  --make-bed\n",
      "  --out coriell_AJ-updated-chr16\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "44056 out of 1278002 variants loaded from .bim file.\n",
      "517 people (323 males, 194 females) loaded from .fam.\n",
      "432 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 517 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.997862.\n",
      "--a1-allele: 44056 assignments made.\n",
      "44056 variants and 517 people pass filters and QC.\n",
      "Among remaining phenotypes, 325 are cases and 107 are controls.  (85 phenotypes\n",
      "are missing.)\n",
      "--make-bed to coriell_AJ-updated-chr16.bed + coriell_AJ-updated-chr16.bim +\n",
      "coriell_AJ-updated-chr16.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_AJ-updated-chr17.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_AJ-HRC.txt\n",
      "  --bfile coriell_AJ-updated\n",
      "  --chr 17\n",
      "  --make-bed\n",
      "  --out coriell_AJ-updated-chr17\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "43174 out of 1278002 variants loaded from .bim file.\n",
      "517 people (323 males, 194 females) loaded from .fam.\n",
      "432 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 517 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.997753.\n",
      "--a1-allele: 43174 assignments made.\n",
      "43174 variants and 517 people pass filters and QC.\n",
      "Among remaining phenotypes, 325 are cases and 107 are controls.  (85 phenotypes\n",
      "are missing.)\n",
      "--make-bed to coriell_AJ-updated-chr17.bed + coriell_AJ-updated-chr17.bim +\n",
      "coriell_AJ-updated-chr17.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_AJ-updated-chr18.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_AJ-HRC.txt\n",
      "  --bfile coriell_AJ-updated\n",
      "  --chr 18\n",
      "  --make-bed\n",
      "  --out coriell_AJ-updated-chr18\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "34995 out of 1278002 variants loaded from .bim file.\n",
      "517 people (323 males, 194 females) loaded from .fam.\n",
      "432 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 517 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998677.\n",
      "--a1-allele: 34995 assignments made.\n",
      "34995 variants and 517 people pass filters and QC.\n",
      "Among remaining phenotypes, 325 are cases and 107 are controls.  (85 phenotypes\n",
      "are missing.)\n",
      "--make-bed to coriell_AJ-updated-chr18.bed + coriell_AJ-updated-chr18.bim +\n",
      "coriell_AJ-updated-chr18.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_AJ-updated-chr19.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_AJ-HRC.txt\n",
      "  --bfile coriell_AJ-updated\n",
      "  --chr 19\n",
      "  --make-bed\n",
      "  --out coriell_AJ-updated-chr19\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "35402 out of 1278002 variants loaded from .bim file.\n",
      "517 people (323 males, 194 females) loaded from .fam.\n",
      "432 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 517 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.997205.\n",
      "--a1-allele: 35402 assignments made.\n",
      "35402 variants and 517 people pass filters and QC.\n",
      "Among remaining phenotypes, 325 are cases and 107 are controls.  (85 phenotypes\n",
      "are missing.)\n",
      "--make-bed to coriell_AJ-updated-chr19.bed + coriell_AJ-updated-chr19.bim +\n",
      "coriell_AJ-updated-chr19.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_AJ-updated-chr20.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_AJ-HRC.txt\n",
      "  --bfile coriell_AJ-updated\n",
      "  --chr 20\n",
      "  --make-bed\n",
      "  --out coriell_AJ-updated-chr20\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "30918 out of 1278002 variants loaded from .bim file.\n",
      "517 people (323 males, 194 females) loaded from .fam.\n",
      "432 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 517 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998091.\n",
      "--a1-allele: 30918 assignments made.\n",
      "30918 variants and 517 people pass filters and QC.\n",
      "Among remaining phenotypes, 325 are cases and 107 are controls.  (85 phenotypes\n",
      "are missing.)\n",
      "--make-bed to coriell_AJ-updated-chr20.bed + coriell_AJ-updated-chr20.bim +\n",
      "coriell_AJ-updated-chr20.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_AJ-updated-chr21.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_AJ-HRC.txt\n",
      "  --bfile coriell_AJ-updated\n",
      "  --chr 21\n",
      "  --make-bed\n",
      "  --out coriell_AJ-updated-chr21\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "17858 out of 1278002 variants loaded from .bim file.\n",
      "517 people (323 males, 194 females) loaded from .fam.\n",
      "432 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 517 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998195.\n",
      "--a1-allele: 17858 assignments made.\n",
      "17858 variants and 517 people pass filters and QC.\n",
      "Among remaining phenotypes, 325 are cases and 107 are controls.  (85 phenotypes\n",
      "are missing.)\n",
      "--make-bed to coriell_AJ-updated-chr21.bed + coriell_AJ-updated-chr21.bim +\n",
      "coriell_AJ-updated-chr21.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_AJ-updated-chr22.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_AJ-HRC.txt\n",
      "  --bfile coriell_AJ-updated\n",
      "  --chr 22\n",
      "  --make-bed\n",
      "  --out coriell_AJ-updated-chr22\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "19266 out of 1278002 variants loaded from .bim file.\n",
      "517 people (323 males, 194 females) loaded from .fam.\n",
      "432 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 517 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.997538.\n",
      "--a1-allele: 19266 assignments made.\n",
      "19266 variants and 517 people pass filters and QC.\n",
      "Among remaining phenotypes, 325 are cases and 107 are controls.  (85 phenotypes\n",
      "are missing.)\n",
      "--make-bed to coriell_AJ-updated-chr22.bed + coriell_AJ-updated-chr22.bim +\n",
      "coriell_AJ-updated-chr22.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to coriell_AJ-updated-chr23.log.\n",
      "Options in effect:\n",
      "  --a1-allele Force-Allele1-coriell_AJ-HRC.txt\n",
      "  --bfile coriell_AJ-updated\n",
      "  --chr 23\n",
      "  --make-bed\n",
      "  --out coriell_AJ-updated-chr23\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "5476 out of 1278002 variants loaded from .bim file.\n",
      "517 people (323 males, 194 females) loaded from .fam.\n",
      "432 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 517 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.997423.\n",
      "--a1-allele: 5476 assignments made.\n",
      "5476 variants and 517 people pass filters and QC.\n",
      "Among remaining phenotypes, 325 are cases and 107 are controls.  (85 phenotypes\n",
      "are missing.)\n",
      "--make-bed to coriell_AJ-updated-chr23.bed + coriell_AJ-updated-chr23.bim +\n",
      "coriell_AJ-updated-chr23.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Executing: plink --bfile coriell_AJ-updated-chr1 --recode vcf --chr 1 --out AJ_chr1\n",
      "Executing: plink --bfile coriell_AJ-updated-chr2 --recode vcf --chr 2 --out AJ_chr2\n",
      "Executing: plink --bfile coriell_AJ-updated-chr3 --recode vcf --chr 3 --out AJ_chr3\n",
      "Executing: plink --bfile coriell_AJ-updated-chr4 --recode vcf --chr 4 --out AJ_chr4\n",
      "Executing: plink --bfile coriell_AJ-updated-chr5 --recode vcf --chr 5 --out AJ_chr5\n",
      "Executing: plink --bfile coriell_AJ-updated-chr6 --recode vcf --chr 6 --out AJ_chr6\n",
      "Executing: plink --bfile coriell_AJ-updated-chr7 --recode vcf --chr 7 --out AJ_chr7\n",
      "Executing: plink --bfile coriell_AJ-updated-chr8 --recode vcf --chr 8 --out AJ_chr8\n",
      "Executing: plink --bfile coriell_AJ-updated-chr9 --recode vcf --chr 9 --out AJ_chr9\n",
      "Executing: plink --bfile coriell_AJ-updated-chr10 --recode vcf --chr 10 --out AJ_chr10\n",
      "Executing: plink --bfile coriell_AJ-updated-chr11 --recode vcf --chr 11 --out AJ_chr11\n",
      "Executing: plink --bfile coriell_AJ-updated-chr12 --recode vcf --chr 12 --out AJ_chr12\n",
      "Executing: plink --bfile coriell_AJ-updated-chr13 --recode vcf --chr 13 --out AJ_chr13\n",
      "Executing: plink --bfile coriell_AJ-updated-chr14 --recode vcf --chr 14 --out AJ_chr14\n",
      "Executing: plink --bfile coriell_AJ-updated-chr15 --recode vcf --chr 15 --out AJ_chr15\n",
      "Executing: plink --bfile coriell_AJ-updated-chr16 --recode vcf --chr 16 --out AJ_chr16\n",
      "Executing: plink --bfile coriell_AJ-updated-chr17 --recode vcf --chr 17 --out AJ_chr17\n",
      "Executing: plink --bfile coriell_AJ-updated-chr18 --recode vcf --chr 18 --out AJ_chr18\n",
      "Executing: plink --bfile coriell_AJ-updated-chr19 --recode vcf --chr 19 --out AJ_chr19\n",
      "Executing: plink --bfile coriell_AJ-updated-chr20 --recode vcf --chr 20 --out AJ_chr20\n",
      "Executing: plink --bfile coriell_AJ-updated-chr21 --recode vcf --chr 21 --out AJ_chr21\n",
      "Executing: plink --bfile coriell_AJ-updated-chr22 --recode vcf --chr 22 --out AJ_chr22\n",
      "Executing: plink --bfile coriell_AJ-updated-chr23 --recode vcf --chr 23 --out AJ_chr23\n",
      "Executing: plink --bfile coriell_AMR --freq --out AMR\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to AMR.log.\n",
      "Options in effect:\n",
      "  --bfile coriell_AMR\n",
      "  --freq\n",
      "  --out AMR\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "1956358 variants loaded from .bim file.\n",
      "131 people (66 males, 65 females) loaded from .fam.\n",
      "116 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 131 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.998249.\n",
      "--freq: Allele frequencies (founders only) written to AMR.frq .\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Executing: perl HRC-1000G-check-bim.pl -b coriell_AMR.bim -f AMR.frq -r PASS.Variantsbravo-dbsnp-all.tab -h\n"
     ]
    }
   ],
   "source": [
    "# run imputation\n",
    "\n",
    "# these files need to be in a different place eventually\n",
    "ref_panel='/data/vitaled2/GenoTools/ref/PASS.Variantsbravo-dbsnp-all.tab'\n",
    "check_bim_pl = '/data/vitaled2/GenoTools/ref/HRC-1000G-check-bim.pl'\n",
    "\n",
    "imputed_out_dir = f'{out_genotypes}/coriell/imputed' \n",
    "os.makedirs(imputed_out_dir, exist_ok=True)\n",
    "\n",
    "# jenky method for pulling cleaned genos with ancestry labels for imputation... should figure out how to do this better later\n",
    "impute_genos_list = [x.split('.')[0] for x in glob.glob(f'{out_dir}/*.bed')]\n",
    "\n",
    "impute_labels_list = [x.split('/')[-1].replace('.bed','').split('_')[-1] for x in impute_genos_list]\n",
    "\n",
    "with open(f'{swarm_scripts_dir}/run_imputation_pipeline.swarm','w') as f:\n",
    "    for geno, label in zip(impute_genos_list, impute_labels_list):\n",
    "\n",
    "        label_temp_outdir = f'{basedir}/coriell/bed/{label}'\n",
    "        label_temp = f'{label_temp_outdir}/{label}'\n",
    "        label_outdir = f'{imputed_out_dir}/{label}'\n",
    "        os.makedirs(f'{label_outdir}', exist_ok=True)\n",
    "        os.makedirs(f'{label_temp_outdir}', exist_ok=True)\n",
    "        impute_data = impute_data_prep(geno, label_temp, ref_panel, check_bim_pl)\n",
    "        run_pipeline = f'python3 ../run_imputation_pipeline.py --geno {label_temp} --token {config.api_key} --out {label_outdir}/'\n",
    "#         run_pipeline = f'python3 ../run_imputation_pipeline.py --geno {geno} --temp {label_temp} --token {config.api_key} --ref_panel {ref_panel} --check_bim_pl {check_bim_pl} --out {label_outdir}/'\n",
    "        f.write(f'{run_pipeline}\\n')\n",
    "        \n",
    "f.close()\n",
    "\n",
    "\n",
    "    \n",
    "# !cat {swarm_scripts_dir}/run_imputation_pipeline.swarm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d36e6d99-be7b-4481-a928-5bc1d97ac45f",
   "metadata": {},
   "outputs": [],
   "source": [
    "!swarm -f {swarm_scripts_dir}/run_imputation_pipeline.swarm -g 20 -t 16 --time=80:00:00 --logdir swarm --gres=lscratch:20 --partition=norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8be260a2-0489-4195-8b50-60e36cdc3b96",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ref_panel='/data/vitaled2/GenoTools/ref/PASS.Variantsbravo-dbsnp-all.tab'\n",
    "check_bim_pl = '/data/vitaled2/GenoTools/ref/HRC-1000G-check-bim.pl'\n",
    "imputed_out_dir = f'{out_genotypes}/coriell/imputed' \n",
    "rerun_geno = '/data/CARD/PD/GP2/genotypes/coriell/clean/coriell_AAC'\n",
    "rerun_label = 'AAC'\n",
    "\n",
    "with open(f'{swarm_scripts_dir}/rerun_imputation_pipeline.swarm','w') as f:\n",
    "\n",
    "    label_temp_outdir = f'{basedir}/coriell/bed/{rerun_label}'\n",
    "    label_temp = f'{label_temp_outdir}/{rerun_label}'\n",
    "    label_outdir = f'{imputed_out_dir}/{rerun_label}'\n",
    "    os.makedirs(f'{label_outdir}', exist_ok=True)\n",
    "    os.makedirs(f'{label_temp_outdir}', exist_ok=True)\n",
    "    impute_data = impute_data_prep(rerun_geno, label_temp, ref_panel, check_bim_pl)\n",
    "    run_pipeline = f'python3 ../run_imputation_pipeline.py --geno {label_temp} --token {config.api_key} --out {label_outdir}/'\n",
    "#         run_pipeline = f'python3 ../run_imputation_pipeline.py --geno {geno} --temp {label_temp} --token {config.api_key} --ref_panel {ref_panel} --check_bim_pl {check_bim_pl} --out {label_outdir}/'\n",
    "    f.write(f'{run_pipeline}\\n')\n",
    "        \n",
    "f.close()\n",
    "\n",
    "!cat {swarm_scripts_dir}/rerun_imputation_pipeline.swarm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2217e9a5-951e-44d3-964a-24d7ec02bb30",
   "metadata": {},
   "outputs": [],
   "source": [
    "!swarm -f {swarm_scripts_dir}/rerun_imputation_pipeline.swarm -g 20 -t 16 --time=80:00:00 --logdir swarm --gres=lscratch:20 --partition=norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "808adf72-05f4-47b5-8810-160eecce876d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05bdb2b2-5ab1-4c46-bba6-c2125646e05c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4aeb078-313f-41a8-9b95-fc1718122b20",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bac1dfe-87d9-47bf-bc3f-73b049b30e2d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68c38965-544d-428a-bd87-58bd05913027",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8287d971-dea0-410b-aa3b-181170f47e8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge AAC with ref panel \n",
    "\n",
    "from QC.utils import merge_genos\n",
    "\n",
    "aac_geno = '/data/CARD/PD/GP2/raw_genotypes/coriell/plink/coriell_callrate_sex_ancestry_AAC_related_het_variant'\n",
    "out_dir = '/data/CARD/PD/GP2/genotypes/coriell/clean'\n",
    "ref_dir_path = '/data/LNG/vitaled2/1kgenomes'\n",
    "ref_panel = f'{ref_dir_path}/1kg_ashkj_ref_panel_gp2_pruned'\n",
    "\n",
    "merged_ref_geno = f'{aac_geno}_merged_ref'\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0dd64a79-699d-4055-af21-f75fa5ecbdac",
   "metadata": {},
   "outputs": [],
   "source": [
    "geno_ancestry_prune1 = f'{aac_geno}_ancestry_prune1'\n",
    "geno_ancestry_prune2 = f'{aac_geno}_ancestry_prune2'\n",
    "# geno_ancestry_prune3 = f'{geno_het}_ancestry_prune3'\n",
    "\n",
    "# prune geno_het for geno, maf, hwe, and palindromes\n",
    "bim = pd.read_csv(f'{aac_geno}.bim', sep='\\t', header=None)\n",
    "\n",
    "# find and drop palindromes in geno_het bim file\n",
    "bim.columns = ['chr', 'rsid', 'kb', 'pos', 'a1', 'a2']\n",
    "palindromes = bim.loc[((bim.a1 == 'A') & (bim.a2 == 'T')) | ((bim.a1 == 'T') & (bim.a2 == 'A')) | ((bim.a1 == 'C') & (bim.a2 == 'G')) | ((bim.a1 == 'G') & (bim.a2 == 'C'))]\n",
    "palindromes['rsid'].to_csv(f'{aac_geno}_palindromes.snplist', header=False, index=False, sep='\\t')\n",
    "\n",
    "plink_cmd1 = f'plink --bfile {aac_geno}\\\n",
    " --maf 0.05\\\n",
    " --geno 0.01\\\n",
    " --hwe 0.0001\\\n",
    " --autosome\\\n",
    " --allow-no-sex\\\n",
    " --exclude {aac_geno}_palindromes.snplist\\\n",
    " --make-bed\\\n",
    " --out {geno_ancestry_prune1}' \n",
    "\n",
    "#remove high-LD regions\n",
    "plink_cmd2 = f'plink --bfile {geno_ancestry_prune1}\\\n",
    " --exclude range {ref_dir_path}/hg19_exclusion_regions.txt\\\n",
    " --autosome\\\n",
    " --allow-no-sex\\\n",
    " --make-bed\\\n",
    " --out {geno_ancestry_prune2}'\n",
    "\n",
    "cmds = [plink_cmd1, plink_cmd2]\n",
    "\n",
    "for cmd in cmds:\n",
    "    shell_do(cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "652419eb-6df6-43f0-bc55-6ecdff2251e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# now get common snps between pruned ref panel and pruned geno\n",
    "geno_pruned_bim = pd.read_csv(f'{geno_ancestry_prune2}.bim', sep='\\t', header=None)\n",
    "geno_pruned_bim.columns = ['chr', 'rsid', 'kb', 'pos', 'a1', 'a2']\n",
    "ref_bim = pd.read_csv(f'{ref_panel}.bim', sep='\\t', header=None)\n",
    "ref_bim.columns = ['chr', 'rsid', 'kb', 'pos', 'a1', 'a2']\n",
    "\n",
    "common_snps = ref_bim.merge(geno_pruned_bim, how='inner', on=['rsid'])\n",
    "\n",
    "common_snps['rsid'].to_csv(f'{out_dir}/aac_geno_ref_panel_common.snps', sep='\\t', header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6901bcea-71a8-4131-92c1-095dc723637f",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_panel_common_snps_geno_out = f'{out_dir}/aac_geno_ref_panel_common_snps'\n",
    "ref_panel_common_snps = f'{out_dir}/aac_geno_ref_panel_common.snps'\n",
    "\n",
    "ext_snps_cmd = f'plink --bfile {ref_panel} --extract {ref_panel_common_snps} --make-bed --out {ref_panel_common_snps_geno_out}'\n",
    "shell_do(ext_snps_cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "362c4d4f-d05c-4eac-b7a9-6e9f589b5139",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get reference alleles from ref_panel_common_snps\n",
    "ref_panel_common_snps_ref_alleles = f'{ref_panel_common_snps_geno_out}.ref_allele'\n",
    "ref_panel_common_snps_bim = pd.read_csv(f'{ref_panel_common_snps_geno_out}.bim', header=None, sep='\\t')\n",
    "ref_panel_common_snps_bim.columns = ['chr', 'rsid', 'kb', 'pos', 'a1', 'a2']\n",
    "ref_panel_common_snps_bim[['rsid','a1']].to_csv(ref_panel_common_snps_ref_alleles, sep='\\t', header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a25b7f7-4351-429b-95f9-1eb665b0500c",
   "metadata": {},
   "outputs": [],
   "source": [
    "geno_common_snps = f'{geno_ancestry_prune2}_common_snps'\n",
    "\n",
    "ext_snps_cmd = f'plink --bfile {geno_ancestry_prune2} --extract {ref_panel_common_snps} --reference-allele {ref_panel_common_snps_ref_alleles} --make-bed --out {geno_common_snps}'\n",
    "shell_do(ext_snps_cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acc34d0b-9ca1-41d5-a9ec-6c8a5407577f",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_ref_geno = f'{geno_ancestry_prune2}_merged_ref'\n",
    "\n",
    "merge_genos(geno_common_snps, ref_panel_common_snps_geno_out, merged_ref_geno)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "044e9712-9d0b-428a-8da9-cc84b86d7b9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "ancestry = pd.read_csv('/data/LNG/vitaled2/1kgenomes/ref_panel_ancestry.txt', sep='\\t', header=None, names=['FID','IID','label'])\n",
    "ref_fam = pd.read_csv(f'{ref_panel}.fam', sep=' ', header=None)\n",
    "ref_labeled = ref_fam.merge(ancestry, how='left', left_on=[0,1], right_on=['FID','IID'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7771a59c-2a56-481d-954e-0830fbfad381",
   "metadata": {},
   "outputs": [],
   "source": [
    "geno_common_snps_df = pd.read_csv(f'{geno_common_snps}.fam', header=None, sep='\\s+', usecols=[0,1], names=['FID','IID'])\n",
    "geno_common_snps_df.loc[:,'label'] = 'new'\n",
    "fs_labels = geno_common_snps_df.append(ref_labeled.loc[:,['FID','IID','label']])\n",
    "fs_labels.to_csv(f'{merged_ref_geno}.labels', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1257371b-52b0-4f9e-9ec1-96699d281ea0",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "ancestry_labels = f'{merged_ref_geno}.labels'\n",
    "structure_out = f'{merged_ref_geno}_structure'\n",
    "# fam = pd.read_csv(f'{ref_panel_prune_final}.fam', sep=' ', header=None)\n",
    "structure = f'/data/vitaled2/ref_panel/fastStructure/structure.py'\n",
    "\n",
    "\n",
    "# run for k=1-8 and use chooseK.py to select best\n",
    "fs_swarm_script = f'{out_dir}/faststructure_choosek.swarm'\n",
    "with open(fs_swarm_script, 'w') as f:\n",
    "    k=8\n",
    "    fs_cmd = f'bash /data/vitaled2/GWAS/gwas/faststructure_setup_and_run.sh -i {merged_ref_geno} -o {structure_out} -f {structure} -k {k}'\n",
    "    f.write(f'{fs_cmd}\\n')\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8624d45-e2b2-4df4-bab2-bbc33c31e78a",
   "metadata": {},
   "outputs": [],
   "source": [
    "shell_do(f'swarm -f {fs_swarm_script} -g 16 --time=10:00:00 -t 20 --logdir {ref_dir_path}/swarm --partition=norm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2faa0352-ac8b-48e8-9eda-90c55e12071b",
   "metadata": {},
   "outputs": [],
   "source": [
    "fam = pd.read_csv(f'{merged_ref_geno}.fam', sep=' ', header=None)\n",
    "pop = pd.read_csv(f'{merged_ref_geno}.labels', sep='\\t')\n",
    "q_df = pd.read_csv(f'{structure_out}.8.meanQ', header=None, sep='\\s+')\n",
    "q_df.columns = [f'pop{i}' for i in range(len(q_df.columns))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5f1c58d-17c2-4530-aa0a-6675d76e83a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "q_df['FID'], q_df['IID'] = fam[0], fam[1]\n",
    "q_pop_merged = q_df.merge(pop, left_on=['FID','IID'], right_on=['FID','IID'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d092753-a588-4c4c-b709-4e08a8292b7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "q_pop_merged.to_csv(f'{out_dir}/aac_samples_labeled_faststructure.txt', sep='\\t', header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "081ed410-5bee-4d4d-928a-832e51ea6597",
   "metadata": {},
   "outputs": [],
   "source": [
    "f'{out_dir}/aac_samples_labeled_faststructure.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a60ff86c-58f8-4250-9600-3e81d32c9e83",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee2f95d0-2073-457e-a66f-3dfeb47eae52",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python/3.7",
   "language": "python",
   "name": "py3.7"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
